---
title: "FastSmart - Figures"
author: "V. Hahaut, D. Pavlinic & S. Picelli."
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r Prerequisits, echo = FALSE, include = TRUE, warning = FALSE, comment = FALSE, message = FALSE, cache = FALSE}

options(scipen=999)

# 1. LIBRARIES / FUNCTIONS
library(SummarizedExperiment)
library(tidyverse)
library(cowplot)

# 2. ANNOTATIONS
gtf.path <- "/home/vincent.hahaut/data_storage/REFERENCES/hsap_fastsmart/REFERENCE/GTF/gencode.v34.primary_assembly.annotation.gtf"
gtf <- rtracklayer::import(gtf.path)
codingIDs <- gtf$gene_name[gtf$gene_type == "protein_coding"]

# 3. COLORS
mycolors <- as.vector(c(RColorBrewer::brewer.pal(name = "Dark2", 8),
                        yarrr::piratepal("basel"), 
                        yarrr::piratepal("nemo"), 
                        yarrr::piratepal("appletv"), 
                        yarrr::piratepal("pony"), 
                        yarrr::piratepal("info"),
                        yarrr::piratepal("decision")[-4],
                        yarrr::piratepal("ipod")))

methodPalette <- function(){scale_fill_manual(values = c("FS" = "#E41A1C",
                                                         "SS2" = "#377EB8",
                                                         "SS3" = "#4DAF4A",
                                                         "SSsc" = "#984EA3")
)}

```

```{r summaryPlots function, echo = FALSE, include = TRUE, warning = FALSE, comment = FALSE, message = FALSE, cache = FALSE}

summaryPlots <- function(mydata = mydata, title = NULL, mycols = NULL){

  if(length(unique(mydata$NAME)) > 2){
    # Kruskall/Dunn's test
    if(broom::tidy(kruskal.test(mydata$count_0, mydata$NAME))$p.value < 0.05){
      mytest <- FSA::dunnTest(mydata$count_0, mydata$NAME, method = "bonferroni")
      mytest <- data_frame(group1 = str_split(mytest$res[,1], " - ", simplify = TRUE)[,1], 
                           group2 =  str_split(mytest$res[,1], " - ", simplify = TRUE)[,2],
                           p.value = mytest$res$P.adj)
    } else {
      mytest <- data_frame(group1 = "A",
                           group2 =  "B",
                           p.value = 1)
    }
  } else {
    # Mann Whitney
    mytest <- broom::tidy(wilcox.test(data = mydata, mydata$count_0 ~ mydata$NAME, alternative = "two.sided"))
    mytest <- data_frame(group1 = levels(factor(mydata$NAME))[1], 
                         group2 =  levels(factor(mydata$NAME))[2], 
                         p.value = mytest$p.value)
  }
  p.gene <- ggplot() +
  geom_hline(yintercept = seq(0,max(mydata$count_0)*1.1,1000), linetype = "dotted", color = "darkgrey") +
  geom_violin(data = mydata, aes(y = count_0, x = NAME, fill = reorder(NAME, count_0, median))) +
  geom_boxplot(data = mydata, aes(y = count_0, x = NAME, fill = reorder(NAME, count_0, median)), width=0.1) +
  scale_fill_manual(values = mycols) +
  scale_y_continuous(breaks = seq(0,max(mydata$count_0)*1.1,1000)) +
  scale_color_brewer(palette = "Dark2") +
  ylab("Detected Genes") + 
  xlab("") +
  theme_cowplot(font_size = 28) +
  theme(legend.position = "none", 
        legend.title = element_blank(),
        panel.background = element_rect(color = "white"),
        plot.background = element_rect(color = "white"),
        axis.text.y = element_text(size = 22), 
        axis.text.x = element_text(size = 20),
        axis.title = element_text(size = 24),
        legend.text = element_text(size = 18)) +
  geom_text(data = mydata, aes(y = 0, x = NAME, label=paste0("n = ", ..count..)), y=0, stat='count', size=6)

  mytest <- filter(mytest, p.value < 0.05)
  if(nrow(mytest) > 0){
    index <- 1:nrow(mytest)
      
    pos.stats <- mytest %>% mutate(
      y.seg = max(mydata$count_0) + max(mydata$count_0)*(1.2*index/50),
      y.lab = y.seg * 1.0005,
      lab = case_when(p.value < 0.05 & p.value > 0.005 ~ "*",
                      p.value < 0.005 & p.value > 0.0005 ~ "**",
                      p.value < 0.0005 ~ "***",
                      TRUE ~ "NA")) %>%
      rowwise() %>%
      mutate(lab.pos = mean(c(which(levels(factor(mydata$NAME)) == group1), which(levels(factor(mydata$NAME)) == group2)))) %>%
      ungroup()

    p.gene <- p.gene +
      geom_segment(data = pos.stats, aes(y = y.seg, yend = y.seg, x = group1, xend = group2), size = 0.25) +
      geom_text(data = pos.stats, aes(y = y.lab, label = lab, x = lab.pos), size = 6)

  }

  p.resqc <- filter(resqc, ID %in% mydata$ID) %>%
  select(ID, UTR3_Exons, UTR5_Exons, CDS_Exons, Introns, TES_down_10kb, TSS_up_10kb, totalReadTags) %>%
  mutate(Intergenic = totalReadTags - UTR3_Exons - UTR5_Exons - CDS_Exons - Introns - TES_down_10kb - TSS_up_10kb) %>%
  gather(key, val, -ID, -totalReadTags) %>%
  mutate(perc = 100*val/totalReadTags) %>%
  left_join(mydata %>% select(NAME, ID), by = "ID") %>%
  mutate(key = case_when(key == "UTR5_Exons" ~ "5'UTR Exons",
                         key == "UTR3_Exons" ~ "3'UTR Exons",
                         key == "CDS_Exons" ~ "CDS Exons",
                         key == "Introns" ~ "Introns",
                         key == "Intergenic" ~ "Intergenic",
                         key == "TES_down_10kb" ~ "TES_down_10kb",
                         key == "TSS_up_10kb" ~ "TSS_up_10kb"),
         key = factor(as.factor(key), levels = c("TES_down_10kb", "TSS_up_10kb", "3'UTR Exons", "5'UTR Exons", "Intergenic", "Introns", "CDS Exons"))) %>%
  # Bring little info to the picture
  filter(!key %in% c("TES_down_10kb", "TSS_up_10kb")) %>%
  ggplot(aes(y = NAME, x = perc, fill = NAME)) +
  geom_violin() +
  geom_boxplot(width = 0.1) +
  theme_cowplot(font_size = 28) +
  facet_wrap("key", scales = "free_x", nrow = 1) +
  scale_fill_manual(values = mycols) +
  theme(axis.text.x = element_text(size = 20),
        legend.position = "none",
        strip.text = element_text(size = 24, face = "bold"),
        panel.background = element_rect(color = "white"),
        plot.background = element_rect(color = "white"),
        rect = element_rect(size = 0.75, color = "black"), 
        panel.border = element_rect(size = 0.75, color = "black"), 
        panel.grid.major.x = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
        axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
  ylab("") +
  xlab("Read Tags (%)") 
  
  
  p.star <- select(mydata, NAME, P_uniquely_mapped_reads, P_multiple_Loci, P_too_short_read) %>%
  gather(key, val, -NAME) %>%
  mutate(key = case_when(key == "P_uniquely_mapped_reads" ~ "Uniquely Mapped", 
                         key == "P_multiple_Loci" ~ "Multi-Mapped", 
                         key == "P_too_short_read" ~ "Unmapped"),
         key = factor(as.factor(key), levels = c("Uniquely Mapped", "Multi-Mapped", "Unmapped")),
         FILL = paste(key, NAME)) %>%
  mutate(val = as.numeric(as.character(val))) %>%
  ggplot(aes(x = NAME, y = val, fill = NAME)) +
  geom_violin() +
  geom_boxplot(width = 0.1) +
  ylab("Reads (%)") +
  xlab("") +
  facet_wrap("key", nrow = 1, scales = "free_y") +
  theme_cowplot(font_size = 28) +
  theme(legend.position = "none",
        panel.background = element_rect(color = "white"),
        plot.background = element_rect(color = "white"),
        rect = element_rect(size = 0.75, color = "black"), 
        strip.text = element_text(size = 18, face = "bold"),
        panel.border = element_rect(size = 0.75, color = "black"), 
        axis.text.x = element_text(size = 20, angle = 45, hjust = 1),
        panel.grid.major.y = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
        axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
  scale_fill_manual(values = mycols) 

  ggsave(bg = "white",
    plot_grid(
      plot_grid(p.gene, p.star, 
                nrow = 1, labels = c("a", "b"), label_size = 28, rel_widths = c(0.32, 0.67)), 
      p.resqc, nrow = 2, label_size = 28, labels = c("", "c")) + theme(panel.background = element_rect(color = "white"),
        plot.background = element_rect(color = "white")), 
  filename = paste0("/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/", title, ".tiff"),
  dpi = 300, width = 32, height = 20)
  
  return(list(p.gene, p.star,p.resqc))
}
summaryPlots <- function(mydata = mydata, title = NULL, mycols = NULL){

  if(length(unique(mydata$NAME)) > 2){
    # Kruskall/Dunn's test
    if(broom::tidy(kruskal.test(mydata$count_0, mydata$NAME))$p.value < 0.05){
      mytest <- FSA::dunnTest(mydata$count_0, mydata$NAME, method = "bonferroni")
      mytest <- data_frame(group1 = str_split(mytest$res[,1], " - ", simplify = TRUE)[,1], 
                           group2 =  str_split(mytest$res[,1], " - ", simplify = TRUE)[,2],
                           p.value = mytest$res$P.adj)
    } else {
      mytest <- data_frame(group1 = "A",
                           group2 =  "B",
                           p.value = 1)
    }
  } else {
    # Mann Whitney
    mytest <- broom::tidy(wilcox.test(data = mydata, mydata$count_0 ~ mydata$NAME, alternative = "two.sided"))
    mytest <- data_frame(group1 = levels(factor(mydata$NAME))[1], 
                         group2 =  levels(factor(mydata$NAME))[2], 
                         p.value = mytest$p.value)
  }
  p.gene <- ggplot() +
  geom_hline(yintercept = seq(0,max(mydata$count_0)*1.1,1000), linetype = "dotted", color = "darkgrey") +
  geom_violin(data = mydata, aes(y = count_0, x = NAME, fill = reorder(NAME, count_0, median))) +
  geom_boxplot(data = mydata, aes(y = count_0, x = NAME, fill = reorder(NAME, count_0, median)), width=0.1) +
  scale_fill_manual(values = mycols) +
  scale_y_continuous(breaks = seq(0,max(mydata$count_0)*1.1,1000)) +
  scale_color_brewer(palette = "Dark2") +
  ylab("Detected Genes") + 
  xlab("") +
  theme_cowplot(font_size = 28) +
  theme(legend.position = "none", 
        legend.title = element_blank(),
        panel.background = element_rect(color = "white"),
        plot.background = element_rect(color = "white"),
        axis.text.y = element_text(size = 22), 
        axis.text.x = element_text(size = 20),
        axis.title = element_text(size = 24),
        legend.text = element_text(size = 18)) +
  geom_text(data = mydata, aes(y = 0, x = NAME, label=paste0("n = ", ..count..)), y=0, stat='count', size=6)

  mytest <- filter(mytest, p.value < 0.05)
  if(nrow(mytest) > 0){
    index <- 1:nrow(mytest)
      
    pos.stats <- mytest %>% mutate(
      y.seg = max(mydata$count_0) + max(mydata$count_0)*(1.2*index/50),
      y.lab = y.seg * 1.0005,
      lab = case_when(p.value < 0.05 & p.value > 0.005 ~ "*",
                      p.value < 0.005 & p.value > 0.0005 ~ "**",
                      p.value < 0.0005 ~ "***",
                      TRUE ~ "NA")) %>%
      rowwise() %>%
      mutate(lab.pos = mean(c(which(levels(factor(mydata$NAME)) == group1), which(levels(factor(mydata$NAME)) == group2)))) %>%
      ungroup()

    p.gene <- p.gene +
      geom_segment(data = pos.stats, aes(y = y.seg, yend = y.seg, x = group1, xend = group2), size = 0.25) +
      geom_text(data = pos.stats, aes(y = y.lab, label = lab, x = lab.pos), size = 6)

  }

  p.resqc <- filter(resqc, ID %in% mydata$ID) %>%
  select(ID, UTR3_Exons, UTR5_Exons, CDS_Exons, Introns, TES_down_10kb, TSS_up_10kb, totalReadTags) %>%
  mutate(Intergenic = totalReadTags - UTR3_Exons - UTR5_Exons - CDS_Exons - Introns - TES_down_10kb - TSS_up_10kb) %>%
  gather(key, val, -ID, -totalReadTags) %>%
  mutate(perc = 100*val/totalReadTags) %>%
  left_join(mydata %>% select(NAME, ID), by = "ID") %>%
  mutate(key = case_when(key == "UTR5_Exons" ~ "5'UTR Exons",
                         key == "UTR3_Exons" ~ "3'UTR Exons",
                         key == "CDS_Exons" ~ "CDS Exons",
                         key == "Introns" ~ "Introns",
                         key == "Intergenic" ~ "Intergenic",
                         key == "TES_down_10kb" ~ "TES_down_10kb",
                         key == "TSS_up_10kb" ~ "TSS_up_10kb"),
         key = factor(as.factor(key), levels = c("TES_down_10kb", "TSS_up_10kb", "3'UTR Exons", "5'UTR Exons", "Intergenic", "Introns", "CDS Exons"))) %>%
  # Bring little info to the picture
  filter(!key %in% c("TES_down_10kb", "TSS_up_10kb")) %>%
  ggplot(aes(y = NAME, x = perc, fill = NAME)) +
  geom_violin() +
  geom_boxplot(width = 0.1) +
  theme_cowplot(font_size = 28) +
  facet_wrap("key", scales = "free_x", nrow = 1) +
  scale_fill_manual(values = mycols) +
  theme(axis.text.x = element_text(size = 20),
        legend.position = "none",
        strip.text = element_text(size = 24, face = "bold"),
        panel.background = element_rect(color = "white"),
        plot.background = element_rect(color = "white"),
        rect = element_rect(size = 0.75, color = "black"), 
        panel.border = element_rect(size = 0.75, color = "black"), 
        panel.grid.major.x = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
        axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
  ylab("") +
  xlab("Read Tags (%)") 
  
  
  p.star <- select(mydata, NAME, P_uniquely_mapped_reads, P_multiple_Loci, P_too_short_read) %>%
  gather(key, val, -NAME) %>%
  mutate(key = case_when(key == "P_uniquely_mapped_reads" ~ "Uniquely Mapped", 
                         key == "P_multiple_Loci" ~ "Multi-Mapped", 
                         key == "P_too_short_read" ~ "Unmapped"),
         key = factor(as.factor(key), levels = c("Uniquely Mapped", "Multi-Mapped", "Unmapped")),
         FILL = paste(key, NAME)) %>%
  mutate(val = as.numeric(as.character(val))) %>%
  ggplot(aes(x = NAME, y = val, fill = NAME)) +
  geom_violin() +
  geom_boxplot(width = 0.1) +
  ylab("Reads (%)") +
  xlab("") +
  facet_wrap("key", nrow = 1, scales = "free_y") +
  theme_cowplot(font_size = 28) +
  theme(legend.position = "none",
        panel.background = element_rect(color = "white"),
        plot.background = element_rect(color = "white"),
        rect = element_rect(size = 0.75, color = "black"), 
        strip.text = element_text(size = 18, face = "bold"),
        panel.border = element_rect(size = 0.75, color = "black"), 
        axis.text.x = element_text(size = 20, angle = 45, hjust = 1),
        panel.grid.major.y = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
        axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
  scale_fill_manual(values = mycols) 

  ggsave(bg = "white",
    plot_grid(
      plot_grid(p.gene, p.star, 
                nrow = 1, labels = c("a", "b"), label_size = 28, rel_widths = c(0.32, 0.67)), 
      p.resqc, nrow = 2, label_size = 28, labels = c("", "c")) + theme(panel.background = element_rect(color = "white"),
        plot.background = element_rect(color = "white")), 
  filename = paste0("/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/", title, ".tiff"),
  dpi = 300, width = 32, height = 20)
  
  return(list(p.gene, p.star,p.resqc))
}
```

```{r Load Data, echo = FALSE, include = TRUE, warning = FALSE, comment = FALSE, message = FALSE, cache = FALSE}

SAMPLESHEET <- read.delim("/home/vincent.hahaut/Desktop/GITHUB/IOB/SampleSheet/FS_GLOBAL_SAMPLESHEET_17112021.txt", sep = "\t")

# Function to preprocess a featurecounts data_frame
# Remove low quality cells, add statistics and cell info
processExpressGenes <- function(mydata = NULL, SAMPLESHEET = SAMPLESHEET, STAR.log = STAR.log){
  mydata %>%
    mutate(oldID = ID,
           ID = str_replace(ID, "^X", ""),
         DEPTH = str_extract(ID, "_5000_|10000_|_20000_|_40000_|_50000_|_75000_|_100000_|_125000_|_250000_|_375000_|_500000_|_625000_|_750000_|_1000000_|_1250000_|_1500000_"),
         DEPTH = str_replace_all(DEPTH, "_", ""),
         TIMES = str_extract(ID, ".$"),
         ID = str_replace(ID, 
                          paste0("_", c(5000,10000,20000,40000,50000,75000,100000,125000,250000,
                                        375000,500000,625000,750000,1000000,1250000,1500000), "_1", collapse = "|"), ""),
         ID = str_replace_all(ID, "\\.", "-")) %>%
    left_join(SAMPLESHEET, by = c("ID" = "FASTQ_ID")) %>%
    left_join(STAR.log, by = "ID") %>%
    filter(as.numeric(N_raw_reads) > 100000, as.numeric(N_uniquely_mapped_reads) > 50000, !is.na(CELL)) %>%
    return()
}

# 1. STAR - Statistics
STAR.log <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/STAR.logs.2021-12-05.rds.bz")

# 2. MEDIAN GENE COUNTS - FEATURECOUNTS - DOWNSAMPLINGS - EXON
med.exon <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/downsampling_medianGenes_featureCounts.exon.all.2021-12-05.rds.bz") %>%
  processExpressGenes(mydata = ., SAMPLESHEET = SAMPLESHEET, STAR.log = STAR.log)

# 3. MEDIAN GENE COUNTS - FEATURECOUNTS - DOWNSAMPLINGS - GENES
med.gene <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/downsampling_medianGenes_featureCounts.genes.all.2021-11-13.rds.bz") %>%
  processExpressGenes(mydata = ., SAMPLESHEET = SAMPLESHEET, STAR.log = STAR.log)

# 4. RAW SCE OBJECTS
sce <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/downsampling_featureCounts.exon.all.2021-12-05.rds.bz")

# 5. Salmon objects
sce.sal.500K.med <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/downsampling_medianGenes_salmon.all.500000.2021-12-05.rds.bz")
sce.sal.250K.med <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/downsampling_medianGenes_salmon.all.250000.2021-12-05.rds.bz")

# 6. RESQC - READ DISTRIBUTION
resqc <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/ReSQC.distribution.2021-12-05.rds.bz")

# 7. CHR-WIDE READ DISTRIBUTIONS
mito <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/mitochondrial.reads. 2021-12-05 .rds.bz")

# 8. RESQC - GENE BODY COVERAGES
cover <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/ReSQC.coverage.2021-12-05.rds.bz")

# 9. Estimated FS Timing
timing <- read_tsv("/home/vincent.hahaut/data_storage/FS_intermediate_file/FS_timing.txt")

# 10. GC content
GC.log <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/GC.logs.2021-12-05.rds.bz")

```

# 1. dCTP

```{r Suppl. dCTP, echo = FALSE, include = TRUE, warning = FALSE, comment = FALSE, message = FALSE, cache = FALSE}

# 1. cDNA Yield

# 1.1. Load data
yield <- read_tsv("/home/vincent.hahaut/data_storage/FS_intermediate_file/HEK_cDNAYield_dCTP.txt", col_names = TRUE) %>% 
  mutate(dCTP = ifelse(dCTP != "no dCTP", str_replace(dCTP, " dCTP", ""), "no dCTP"),
         dCTP = factor(dCTP, levels = c("no dCTP", "1.65 mM", "3.3 mM", "5.5 mM", "8.25 mM", "11 mM"))) %>%
  filter(!is.na(Yield))

# 1.2. Compare groups
# Normality is not guarantied (shapiro)
mytest <- FSA::dunnTest(yield$Yield, yield$dCTP, method = "bonferroni")
mytest <- data_frame(group1 = str_split(mytest$res[,1], " - ", simplify = TRUE)[,1], 
                           group2 =  str_split(mytest$res[,1], " - ", simplify = TRUE)[,2],
                           p.value = mytest$res$P.adj) %>%
  filter(p.value < 0.05)

index <- 1:nrow(mytest)
      
pos.stats <- mytest %>% 
  mutate(group1 = factor(group1, levels = c("no dCTP", "1.65 mM", "3.3 mM", "5.5 mM", "8.25 mM", "11 mM")),
         group2 = factor(group2, levels = c("no dCTP", "1.65 mM", "3.3 mM", "5.5 mM", "8.25 mM", "11 mM"))) %>%
  mutate(
  y.seg = max(yield$Yield) + max(yield$Yield)*(1.2*index/50),
  y.lab = y.seg * 1.0005,
  lab = case_when(p.value < 0.05 & p.value > 0.005 ~ "*",
                  p.value < 0.005 & p.value > 0.0005 ~ "**",
                  p.value < 0.0005 ~ "***",
                  TRUE ~ "NA")) %>%
  arrange(group2) %>%
  rowwise() %>%
  mutate(lab.pos = mean(c(which(levels(factor(yield$dCTP)) == group1), which(levels(factor(yield$dCTP)) == group2)))) %>%
  ungroup()

# 1.3. Visualize
pA <- ggplot() +
  geom_hline(yintercept = seq(0,30,5), linetype = "dotted", color = "darkgrey") + 
  geom_boxplot(data = yield, aes(x = dCTP, y = Yield, fill = dCTP), width = 0.8) +
  theme_cowplot(font_size = 24) +
  theme(legend.position = "none", 
        panel.background = element_rect(fill = "white"),
        plot.background = element_rect(fill = "white")) +
  xlab("Extra dCTP") +
  ylab("cDNA Yield (ng/µL)") +
  scale_fill_manual(values = c(
  "no dCTP" =  "#FFFFCC",
  "1.65 mM" = "#A1DAB4",
  "3.3 mM" = "#E41A1C",  
  "5.5 mM" = "#41B6C4",
  "8.25 mM" = "#2C7FB8",
  "11 mM" = "#253494")) +
  geom_text(data = data.frame(), aes(x = levels(yield$dCTP) , y = 0, 
                                     label = paste("n =", table(yield$dCTP)))) +
  geom_segment(data = pos.stats, aes(y = y.seg, yend = y.seg, x = group1, xend = group2), size = 0.25) +
  geom_text(data = pos.stats, aes(y = y.lab, label = lab, x = lab.pos), size = 7)

ggsave(pA, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_cDNA_yield_dCTP.tiff", dpi = 300, height = 7, width = 12)

# 2. Summary Plot
mydata <- med.exon %>%
  filter(PLATE == "309") %>%
  mutate(NAME = paste0(dCTP, " mM"),
         NAME = ifelse(NAME == "0 mM", "no dCTP", NAME),
         NAME = factor(NAME, levels = c("no dCTP", "1.65 mM", "3.3 mM", "5.5 mM", "8.25 mM", "11 mM"))) %>%
  filter(!is.na(NAME), DEPTH == "250000")
write.table(mydata$ID, "/home/vincent.hahaut/Desktop/IDs.txt", col.names = F, row.names = F, quote = F, append = TRUE)

summaryPlots(mydata = mydata, title = "dCTP", mycols = c(
  "no dCTP" =  "#FFFFCC",
  "1.65 mM" = "#A1DAB4",
  "3.3 mM" = "#E41A1C",  
  "5.5 mM" = "#41B6C4",
  "8.25 mM" = "#2C7FB8",
  "11 mM" = "#253494"))

# 3. GC Content

# 3.1. Load the gene GC info
gc.perc <- read.table(file="/home/vincent.hahaut/data_storage/REFERENCES/hsap_fastsmart/REFERENCE/GTF/gencode.v34.exon.GC_lengths.tsv", header = T, sep="\t")

# 3.2. Get the gene expressed in at least 3 cells
HEK.expressed.genes <- row.names(sce.all)[rowSums(counts(sce.all[,colnames(sce.all) %in% paste0("X", mydata$ID, "_250000_1")]) > 0) >= 3]

# 3.3. Mean cell GC content
gc.mean <- mean(100*gc.perc$GC[gc.perc$gene_name %in% HEK.expressed.genes])
 
# 3.4. Aggregate results
gc.perc <- GC.log %>%
    right_join(
        select(mydata, ID, NAME), by = "ID") %>%
  group_by(ID) %>%
  mutate(total = sum(read_count),
         perc = 100*read_count/unique(total)) 
  
# 3.5. Visualize
GC.proportion <- ggplot() +
  geom_hline(yintercept = 0, linetype = "dotted") + 
  geom_vline(xintercept = gc.mean, linetype = "dashed", color = "red") + 
  geom_smooth(data = gc.perc, aes(y = perc, x = `GC%`, color = NAME, fill = NAME), 
              size = 1.5, alpha = 0.05) +
    theme_cowplot(font_size = 24) +
    theme(legend.title = element_blank(),
          panel.background = element_rect(fill = "white"),
          plot.background = element_rect(fill = "white")) +
  xlab("GC (%)") +
  ylab("Mapped Reads (%)") +
  scale_fill_manual(values = c(
  "no dCTP" =  "#FFFFCC",
  "1.65 mM" = "#A1DAB4",
  "3.3 mM" = "#E41A1C",  
  "5.5 mM" = "#41B6C4",
  "8.25 mM" = "#2C7FB8",
  "11 mM" = "#253494")) +
  scale_color_manual(values = c(
  "no dCTP" =  colorspace::darken("#FFFFCC", 0.2),
  "1.65 mM" =  colorspace::darken("#A1DAB4", 0.2),
  "3.3 mM" =  colorspace::darken("#E41A1C", 0.2), 
  "5.5 mM" =  colorspace::darken("#41B6C4", 0.2),
  "8.25 mM" =  colorspace::darken("#2C7FB8", 0.2),
  "11 mM" =  colorspace::darken("#253494", 0.2))) 

ggsave(GC.proportion, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_GCprop.tiff", dpi = 300, height = 7, width = 12, bg = "white")

# 4. Differential Expression

# 4.1. Get the counts from the cell of interest
HEK.GC.sce <- sce.all[,colnames(sce.all) %in% mydata$oldID]

# 4.2. Normalize count matrix (log + size factor)
lib.sf <- scater::librarySizeFactors(HEK.GC.sce)
HEK.GC.sce <- scater::logNormCounts(HEK.GC.sce, size_factors=lib.sf)

# 4.3. Add groups
HEK.GC.sce$GROUP <- str_replace(mydata$NAME, "\\.|\\ ", "_")[match(colnames(HEK.GC.sce), mydata$oldID)]

# 4.4. Filter out low quality cells
HEK.GC.sce <- HEK.GC.sce[rowSums(counts(HEK.GC.sce) > 0) >= 5,]
DE <- scran::findMarkers(HEK.GC.sce, test.type = "t", groups = HEK.GC.sce$GROUP)

# 4.5. Get the "no dCTP" as reference
DE.nodCTP <- DE[[6]]

# 4.6. Aggregate results
DE.nodCTP <- DE.nodCTP %>%
  as.data.frame() %>%
    rownames_to_column("gene_name") %>%
    left_join(gc.perc, by = "gene_name") %>%
    select(-Top, -p.value, -FDR, -summary.logFC) %>%
  pivot_longer(cols = c("logFC.1_65.mM", "logFC.11_mM", "logFC.3_3.mM", "logFC.5_5.mM", "logFC.8_25.mM"), 
               values_to = "logFold") %>%
  mutate(CATEGORY = ifelse(logFold < -1, "DE", "no DE"),
         GC = 100*GC) %>%
  mutate(name = case_when(
    name == "logFC.1_65.mM" ~ "1.65 mM", 
    name == "logFC.11_mM" ~ "11 mM", 
    name == "logFC.3_3.mM" ~ "3.3 mM", 
    name == "logFC.5_5.mM" ~ "5.5 mM", 
    name == "logFC.8_25.mM" ~ "8.25 mM")) %>%
  # Add the GC percentage of the expressed genes in no dCTP condition
  mutate(name = factor(name, levels = c("no dCTP", "1.65 mM", "3.3 mM", "5.5 mM", "8.25 mM", "11 mM")))

# 4.7. Compare groups
# paired wilcoxon
ggpubr::compare_means(GC ~ CATEGORY, group.by = "name", data = DE.nodCTP, p.adjust.method = "bonferroni") 

# 4.8. Visualize results
p.GC.log <- ggplot() +
  geom_hline(yintercept = seq(0,100,10), linetype = "dashed", color = "darkgrey") +
  geom_boxplot(data = DE.nodCTP, aes(x = name, y = GC, fill = CATEGORY), width = 0.5) +
  scale_fill_brewer(palette = "Set1") +
  theme_cowplot(font_size = 24) +
  ylab("GC%") +
  xlab("Extra dCTP") +
  theme(legend.title = element_blank()) +
  geom_segment(data = data_frame(
    y = rep(95, 5),
    x = seq(.8,5,1),
    xend = x+0.4
  ), aes(y = y, yend = y, x = x, xend = xend), size = 0.75) +
  geom_text(data = data_frame(), aes(y = c(97.5,95.5,95.5,95.5,95.5), label = c("ns", "***", "***", "***", "***"), x = seq(1,5,1)), size = 6)

ggsave(p.GC.log, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_GC_DE_dCTP.tiff", dpi = 300, height = 7, width = 12, bg = "white")


# 5. Gene Coverage

# 5.1. Aggregate
mycover <- right_join(cover, mydata %>% select(ID, NAME), by = "ID") %>%
  gather(Percentile, reads, -ID, -NAME) %>%
  filter(reads > 0) %>%
  group_by(ID) %>%
  mutate(percentage = 100*reads / sum(reads)) %>%
  group_by(NAME, Percentile) %>%
  summarise(mean.c = mean(percentage),
            sd.c = sd(percentage)) %>%
  ungroup() 
  
# 5.2. Visualize
p.cover <- ggplot(mycover, aes(x = as.numeric(Percentile), color = as.factor(NAME), fill = as.factor(NAME))) +
  geom_ribbon(aes(ymin = mean.c - sd.c, ymax = mean.c + sd.c), colour = NA) +
  geom_line(aes(y=mean.c), size = 1.25) +
  theme_cowplot(font_size = 24) +
  scale_fill_manual(values = c(
  "no dCTP" =  "#FFFFCC",
  "1.65 mM" = "#A1DAB4",
  "3.3 mM" = "#E41A1C",  
  "5.5 mM" = "#41B6C4",
  "8.25 mM" = "#2C7FB8",
  "11 mM" = "#253494")) +
  scale_color_manual(values = c(
  "no dCTP" =  colorspace::darken("#FFFFCC", 0.2),
  "1.65 mM" =  colorspace::darken("#A1DAB4", 0.2),
  "3.3 mM" =  colorspace::darken("#E41A1C", 0.2), 
  "5.5 mM" =  colorspace::darken("#41B6C4", 0.2),
  "8.25 mM" =  colorspace::darken("#2C7FB8", 0.2),
  "11 mM" =  colorspace::darken("#253494", 0.2))) +
  xlab("Gene Body Percentile (5'->3')") +
  ylab("Percentage Coverage") +
  theme(legend.title = element_blank()) + 
  theme(strip.background = element_blank(),
        strip.text = element_text(face = "bold"))
  
ggsave(p.cover, filename = paste0("/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/dCTP_COVERAGE.tiff"), dpi = 300, height = 6, width = 10, bg = "white")

# 6. Per-base Mismatches

# 6.1. Aggregate
mismatches <- STAR.log %>%
  right_join(
    select(mydata, ID, NAME), by = "ID") %>%
  mutate(P_mismatch = 100*as.numeric(P_mismatch))

# 6.2. Statistical significance
mytest <- FSA::dunnTest(mismatches$P_mismatch, mismatches$NAME, method = "bonferroni")
mytest <- data_frame(group1 = str_split(mytest$res[,1], " - ", simplify = TRUE)[,1], 
                           group2 =  str_split(mytest$res[,1], " - ", simplify = TRUE)[,2],
                           p.value = mytest$res$P.adj) %>%
  filter(p.value < 0.05)

index <- 1:nrow(mytest)
      
pos.stats <- mytest %>% 
  # order or 10x dCTP mess up the order in the plot
  mutate(group1 = factor(group1, levels = c("no dCTP", "1.65 mM", "3.3 mM", "5.5 mM", "8.25 mM", "11 mM")),
         group2 = factor(group2, levels = c("no dCTP", "1.65 mM", "3.3 mM", "5.5 mM", "8.25 mM", "11 mM"))) %>%
  mutate(
  y.seg = max(mismatches$P_mismatch) + max(mismatches$P_mismatch)*(1.2*index/50),
  y.lab = y.seg * 1.0005,
  lab = case_when(p.value < 0.05 & p.value > 0.005 ~ "*",
                  p.value < 0.005 & p.value > 0.0005 ~ "**",
                  p.value < 0.0005 ~ "***",
                  TRUE ~ "NA")) %>%
  arrange(group2) %>%
  rowwise() %>%
  mutate(lab.pos = mean(c(which(levels(factor(mismatches$NAME)) == group1), which(levels(factor(mismatches$NAME)) == group2)))) %>%
  ungroup()

# 6.3. Visualize
p.mismatches <- ggplot() + 
  geom_hline(yintercept = seq(0,60,10), linetype = "dotted", color = "darkgrey") + 
  geom_boxplot(data = mismatches, aes(x = NAME, y = P_mismatch, fill = NAME), width = 0.8) +
  theme_cowplot(font_size = 16) +
  theme(legend.position = "none", 
        panel.background = element_rect(fill = "white"),
        plot.background = element_rect(fill = "white")) +
  xlab("Extra dCTP") +
  ylab("Mismatch Per 100K Mapped Base") +
  scale_fill_manual(values = c(
  "no dCTP" =  "#FFFFCC",
  "1.65 mM" = "#A1DAB4",
  "3.3 mM" = "#E41A1C",  
  "5.5 mM" = "#41B6C4",
  "8.25 mM" = "#2C7FB8",
  "11 mM" = "#253494")) +
  stat_summary(fun.data = function(x){return(data.frame(y = 0, label = paste0("n =", length(x))))}, geom = "text", 
                   position = position_dodge(width = 0.75), size = 5) +
  geom_segment(data = pos.stats, aes(y = y.seg, yend = y.seg, x = group1, xend = group2), size = 0.25) +
  geom_text(data = pos.stats, aes(y = y.lab, label = lab, x = lab.pos), size = 5)

ggsave(p.mismatches, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_mismatches.tiff", dpi = 300, height = 6, width = 8)

# 7. Variant Calling

# 7.1. Load the VCF
vcf <- lapply(list.files("/home/vincent.hahaut/data_storage/210906_NB551561_0058_AH5MLFBGXK/VARIANTS/", pattern = ".GATK.ALL.vcf", recursive = T, full.names = TRUE), function(x) vcfR::read.vcfR(x))
vcf.id <- list.files("/home/vincent.hahaut/data_storage/210906_NB551561_0058_AH5MLFBGXK/VARIANTS/", pattern = ".GATK.ALL.vcf", recursive = TRUE, full.names = FALSE) %>% str_extract(pattern = ".*/") %>% str_replace("\\/STAR\\/", "")

# 7.2 Filter VCF and tidy
vcf <- lapply(seq_along(vcf), function(x)
  vcfR::vcfR2tidy(vcf[[x]], format_fields = c("GT", "DP"))$fix %>%
    mutate(ID = vcf.id[x],
           GROUP = case_when(
             grepl("309_NOdCTP", vcf.id[x]) ~ "no dCTP",
             grepl("309_dCTP1-5x", vcf.id[x]) ~ "1.65 mM",
             grepl("309_dCTP3x", vcf.id[x]) ~ "3.3 mM",
             grepl("309_dCTP5x", vcf.id[x]) ~ "5.5 mM",
             grepl("309_dCTP7-5x", vcf.id[x]) ~ "8.25 mM",
             grepl("309_dCTP10x", vcf.id[x]) ~ "11 mM")))
  
vcf.all <- lapply(vcf, function(x ) x %>% 
                    dplyr::select(ChromKey,CHROM,POS,ID,REF,ALT,QUAL,DP,AF1,GROUP, VDB)) %>% bind_rows %>%
  filter(nchar(REF) == 1 & nchar(ALT) == 1)

# NOT USED ==> too complicated to interpret with the current data
# 7.3. Trinucleotide Context
# gr <- GenomicRanges::makeGRangesFromDataFrame(vcf.all[,c("CHROM", "POS", "REF", "ALT")], keep.extra.columns = TRUE, seqnames.field = "CHROM", start.field = "POS", end.field = "POS")
# vr <- VariantAnnotation::VRanges(seqnames = seqnames(gr), ranges = IRanges(start(gr), end(gr)), ref = gr$REF, alt = gr$ALT)
# vr <- SomaticSignatures::mutationContext(vr, Rsamtools::FaFile("/home/vincent.hahaut/data_storage/REFERENCES/hsap_fastsmart/REFERENCE/DNA/GRCh38.primary_assembly.genome.fa"))
# vr.tidy <- as_tibble(vr) %>% 
#   mutate(REF.tri = str_replace(context, "\\.", ref),
#          ALT.tri = str_replace(context, "\\.", alt))

# vcf.all$REF.tri <- vr.tidy$REF.tri
# vcf.all$ALT.tri <- vr.tidy$ALT.tri
# vcf.all$SUBSTITUTION <- sapply(str_split(vr.tidy$alteration, ""), function(x) paste0(x[1], ">", x[2]))

# 7.4. Compare the % of the different variants
vcf.stats <- vcf.all.tidy %>% 
  mutate(TYPE = ifelse(grepl("->C|->G", CATEGORY), "->C/G", "->A/T"))

ggpubr::compare_means(data = filter(vcf.stats, quality == "QUAL ≥ 20 & DP > 4"), n ~ GROUP, group.by = c("TYPE"), ref.group = "no dCTP")
ggpubr::compare_means(data = filter(vcf.stats, quality != "QUAL ≥ 20 & DP > 4"), n ~ GROUP, group.by = c("TYPE"), ref.group = "no dCTP")

# 7.5. Aggregate results
vcf.all.tidy <- vcf.all %>%
    mutate(CATEGORY = paste0(REF, "->", ALT),
           quality = case_when(
               QUAL >= 20 & DP > 4 ~ "QUAL ≥ 20 & DP > 4",
               TRUE ~ "QUAL < 20"),
           GROUP = factor(GROUP, levels = c("no dCTP", "1.65 mM", "3.3 mM", "5.5 mM", "8.25 mM", "11 mM")),
           quality = factor(quality, levels = c("QUAL ≥ 20 & DP > 4", "QUAL < 20"))) %>%
    filter(!grepl(",", ALT)) %>%
    group_by(ID) %>%
    mutate(total.var = n()) %>%
    ungroup() %>%
    filter(total.var > 100) %>% 
    group_by(ID, CATEGORY, GROUP, quality) %>% 
    summarise(n = 100*n()/unique(total.var)) %>%
    ungroup() %>%
    mutate(CATEGORY=factor(CATEGORY, levels = c("C->G","G->C","A->T","T->A","G->T","C->A","A->C","T->G","T->C","A->G","G->A","C->T")))

# 7.6. Visualize - Percentage of each substitution
p.Cmut <- ggplot(data = vcf.all.tidy, aes(x = GROUP, fill = CATEGORY, y = n)) +
    geom_boxplot(outlier.size = 0.5) +
    theme_bw(base_size = 28) +
    xlab("Extra dCTP") +
    ylab("Substitutions") +
    theme(legend.title = element_blank(), 
          panel.background = element_rect(fill = "white"),
          plot.background = element_rect(fill = "white")) +
    scale_fill_manual(values = mycolors) +
    scale_color_manual(values = mycolors) +
    facet_wrap(~quality, nrow = 2, scales = "free_y")

# 8. Total number of detected variants per 100K reads

# 8.1. Aggregate
vcf.total.tidy <- vcf.all %>% 
  mutate(ID = str_replace(ID, "\\/", "")) %>%
    left_join(STAR.log %>% dplyr::select(ID, N_uniquely_mapped_reads), by = "ID") %>%
    mutate(CATEGORY = paste0(REF, "->", ALT),
           quality = case_when(
               QUAL >= 20 & DP > 4 ~ "QUAL ≥ 20 & DP > 4",
               TRUE ~ "QUAL < 20"),
           GROUP = factor(GROUP, levels = c("no dCTP", "1.65 mM", "3.3 mM", "5.5 mM", "8.25 mM", "11 mM")),
           quality = factor(quality, levels = c("QUAL ≥ 20 & DP > 4", "QUAL < 20"))) %>%
    filter(!grepl(",", ALT)) %>%
    group_by(ID, GROUP, quality) %>%
    mutate(total.var = 100000*n()/as.numeric(N_uniquely_mapped_reads)) %>%
    ungroup() %>%
  dplyr::select(GROUP, quality, total.var) %>% distinct()

# 8.1. Compare groups
vcf.total.tidy %>%
  group_by(quality) %>%
  rstatix::dunn_test(p.adjust.method = "bonferroni", total.var ~ GROUP) %>%
  filter(group1 == "no dCTP")

# 8.2. Visualize
p.Cmut.Total <- ggplot(data = vcf.total.tidy, aes(x = GROUP, fill = quality, y = total.var)) +
    geom_boxplot(outlier.size = 0.5, width = 0.75) +
    theme_bw(base_size = 28) +
    xlab("Extra dCTP") +
    ylab("Substitutions Per 100K\nUniquely Mapped Reads") +
    theme(legend.position = "none", axis.text.x = element_text(angle = 45, hjust = 1),
          panel.background = element_rect(fill = "white"),
          plot.background = element_rect(fill = "white")) +
    scale_fill_manual(values = mycolors) +
    scale_color_manual(values = mycolors) +
    facet_wrap(~quality, nrow = 2, scales = "free_y")

ggsave(p.Cmut.Total, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_Ctransitions.total.tiff", dpi = 300, bg = "white", height = 12, width = 8)



# 9. Trinucleotide Context 
# trinucl <- vcf.all %>%
#     mutate(CATEGORY = paste0(REF.tri, "->", ALT.tri),
#            quality = case_when(
#               QUAL >= 20 & DP > 4 ~ "QUAL ≥ 20 & DP > 4",
#                TRUE ~ "QUAL < 20"),
#            GROUP = factor(GROUP, levels = c("no dCTP", "1.65 mM", "3.3 mM", "5.5 mM", "8.25 mM", "11 mM")),
#            quality = factor(quality, levels = c("QUAL ≥ 20 & DP > 4", "QUAL < 20"))) %>%
#     filter(!grepl(",", ALT)) %>%
#    group_by(ID) %>%
#     mutate(total.var = n()) %>%
#     ungroup() %>%
#     filter(total.var > 100) %>% 
#   group_by(ID, CATEGORY, GROUP, quality, SUBSTITUTION, total.var) %>% 
#     summarise(perc = 100*n()/unique(total.var)) %>%
#     ungroup() %>%
#   group_by(CATEGORY, quality) %>% 
#   mutate(med = median(perc[GROUP == "3.3 mM"]/median(perc[GROUP == "no dCTP"])),
#          n = n()) %>%
#   ungroup() %>% 
#   filter(med > 1.5 | med < 0.50, n > 20)

# p.trinucl <- ggplot(trinucl) +
#     geom_boxplot(aes(x = CATEGORY, fill = GROUP, y = perc)) +
#     theme_bw(base_size = 18) +
#     xlab("Extra dCTP") +
#     ylab("Variants (%)") +
#     theme(legend.title = element_blank(), 
#           panel.background = element_rect(fill = "white"),
#         plot.background = element_rect(fill = "white")) +
#     scale_fill_manual(values = mycolors) +
#     facet_wrap(~quality+SUBSTITUTION, nrow = 2, scales = "free") +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 12))

# ggsave(p.trinucl, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_Ctransitions.tri.tiff", dpi = 300, bg = "white", height = 18, width = 36)


```
