---
  title: "FastSmart - Figures"
author: "V. Hahaut, D. Pavlinic & S. Picelli."
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---
  
```{r Prerequisits, echo = FALSE, include = TRUE, warning = FALSE, comment = FALSE, message = FALSE, cache = FALSE}

options(scipen=999)

# 1. LIBRARIES / FUNCTIONS
library(SummarizedExperiment)
library(tidyverse)
library(cowplot)
source("/home/vincent.hahaut/Desktop/FLASH-Seq/R/general/4_aggregate_files_functions.R")

# 2. ANNOTATIONS
gtf.path <- "/home/vincent.hahaut/data_storage/REFERENCES/hsap_fastsmart/REFERENCE/GTF/gencode.v34.primary_assembly.annotation.gtf"
gtf <- rtracklayer::import(gtf.path)

# 3. COLORS
mycolors <- as.vector(c(RColorBrewer::brewer.pal(name = "Dark2", 8),
                        yarrr::piratepal("basel"), 
                        yarrr::piratepal("nemo"), 
                        yarrr::piratepal("appletv"), 
                        yarrr::piratepal("pony"), 
                        yarrr::piratepal("info"),
                        yarrr::piratepal("decision")[-4],
                        yarrr::piratepal("ipod")))

methodPalette <- function(){scale_fill_manual(values = c("FS" = "#E41A1C",
                                                         "SS2" = "#377EB8",
                                                         "SS3" = "#4DAF4A",
                                                         "SSsc" = "#984EA3")
)}


```

```{r local functions, echo = FALSE, include = TRUE, warning = FALSE, comment = FALSE, message = FALSE, cache = FALSE}

#' processExpressGenes
#' 
#' Merge info from number of genes per cell and STAR mapping statistics
#' First filtering of cells with shallow coverage
#' @param mydata, data.frame generated with sceToExpressedGenes
#' @param SAMPLESHEET, Global samplesheet containing all the details related to the sample preparation
#' @param STAR.log, data.frame generated with getSTARLog
processExpressGenes <- function(mydata = NULL, SAMPLESHEET = SAMPLESHEET, STAR.log = STAR.log){
  depths <- c(5000,10000,20000,40000,50000,75000,100000,125000,250000,375000,500000,625000,750000,1000000,1250000,1500000)
    
  mydata %>%
    mutate(oldID = ID,
           ID = str_replace(ID, "^X", ""),
           DEPTH = str_extract(ID, paste0(paste0("_", depths, "_"), collapse = "|")),
           DEPTH = str_replace_all(DEPTH, "_", ""),
           TIMES = str_extract(ID, ".$"),
           ID = str_replace(ID, paste0("_", depths, "_1", collapse = "|"), ""),
           ID = str_replace_all(ID, "\\.", "-")) %>%
    left_join(SAMPLESHEET, by = c("ID" = "FASTQ_ID")) %>%
    left_join(STAR.log, by = "ID") %>%
    filter(as.numeric(N_raw_reads) > 100000, as.numeric(N_uniquely_mapped_reads) > 50000) %>%
    return()
}


#' summaryPlots
#' 
#' Quick graphic summary of a sequencing condition
#' Returns information regarding 
#' condition are defined in the "NAME" column
#' @param mydata, data.frame generated with sceToExpressedGenes
#' @param title, Suffix of the graph name (save)
#' @param mycols, ggplot scale_color_manual() compatible condition to color (e.g., c("CONDITION_A" = "blue"))

summaryPlots <- function(mydata = mydata, title = NULL, mycols = NULL){
  
  # 1. Gene counts
  
  # 1.1. Statistical test
  # Shapiro.test always reject the normal distribution
  # Kruskall (>2 conditions) or mann-whitney (= 2 conditions)
  if(length(unique(mydata$NAME)) > 2){
    
    if(broom::tidy(kruskal.test(mydata$count_0, mydata$NAME))$p.value < 0.05){
      mytest <- mydata %>%
        rstatix::dunn_test(count_0 ~ NAME, p.adjust.method = "bonferroni")%>%
        rstatix::add_y_position(fun = "max", scales = "free") %>%
        mutate(p.adj.signif = formatC(p, format = "e", digits = 2))
      
      y <- min(diff(mytest$y.position))
      mytest <- mytest %>% 
        filter(p.adj < 0.05) %>%
        mutate(y.position = seq(min(y.position), (min(y.position)+y*(n()-1)), y))
    } else {
      mytest <- data_frame(group1 = "A",
                           group2 =  "B",
                           p.value = 1)
    }
  } else {

      mytest <- mydata %>%
        rstatix::wilcox_test(count_0 ~ NAME) %>%
        rstatix::add_y_position(fun = "max", scales = "free") %>%
        mutate(p.adj.signif = formatC(p, format = "e", digits = 2)) %>%
        filter(p < 0.05)

  }
  
  # 1.2. Display the gene counts
  p.gene <- ggplot() +
    geom_hline(yintercept = seq(0,max(mydata$count_0)*1.1,1000), linetype = "dotted", color = "darkgrey") +
    geom_violin(data = mydata, aes(y = count_0, x = NAME, fill = reorder(NAME, count_0, median))) +
    geom_boxplot(data = mydata, aes(y = count_0, x = NAME, fill = reorder(NAME, count_0, median)), width=0.1) +
    ggpubr::stat_pvalue_manual(mytest, label = "p = {p.adj.signif}", step.increase = 0.05, tip.length = 0.01, size = 6) +
    scale_fill_manual(values = mycols) +
    scale_y_continuous(breaks = seq(0,max(mydata$count_0)*1.1,1000), labels = as.character(formatC(seq(0,max(mydata$count_0)*1.1,1000), format="d", big.mark=",") )) +
    scale_color_brewer(palette = "Dark2") +
    ylab("Detected Genes") + 
    xlab("") +
    theme_cowplot(font_size = 28) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.background = element_rect(color = "white"),
          plot.background = element_rect(color = "white"),
          axis.text.y = element_text(size = 22), 
          axis.text.x = element_text(size = 20),
          axis.title = element_text(size = 24),
          legend.text = element_text(size = 18)) +
    geom_text(data = mydata, aes(y = 0, x = NAME, label=paste0("n = ", ..count..)), y=0, stat='count', size=6)

  
  # 2. ReSQC read tag distribution
  p.resqc <- filter(resqc, ID %in% mydata$ID) %>%
    select(ID, UTR3_Exons, UTR5_Exons, CDS_Exons, Introns, TES_down_10kb, TSS_up_10kb, totalReadTags) %>%
    mutate(Intergenic = totalReadTags - UTR3_Exons - UTR5_Exons - CDS_Exons - Introns) %>%
    gather(key, val, -ID, -totalReadTags) %>%
    mutate(perc = 100*val/totalReadTags) %>%
    left_join(mydata %>% select(NAME, ID), by = "ID") %>%
    mutate(key = case_when(key == "UTR5_Exons" ~ "5'UTR Exons",
                           key == "UTR3_Exons" ~ "3'UTR Exons",
                           key == "CDS_Exons" ~ "CDS Exons",
                           key == "Introns" ~ "Introns",
                           key == "Intergenic" ~ "Intergenic",
                           key == "TES_down_10kb" ~ "TES_down_10kb",
                           key == "TSS_up_10kb" ~ "TSS_up_10kb"),
           key = factor(as.factor(key), levels = c("TES_down_10kb", "TSS_up_10kb", "3'UTR Exons", "5'UTR Exons", "Intergenic", "Introns", "CDS Exons"))) %>%
    # TES / TSS info as they usually bring little info
    filter(!key %in% c("TES_down_10kb", "TSS_up_10kb")) %>%
    ggplot(aes(y = NAME, x = perc, fill = NAME)) +
    geom_violin() +
    geom_boxplot(width = 0.1) +
    theme_cowplot(font_size = 28) +
    facet_wrap("key", scales = "free_x", nrow = 1) +
    scale_fill_manual(values = mycols) +
    theme(axis.text.x = element_text(size = 20),
          legend.position = "none",
          strip.text = element_text(size = 24, face = "bold"),
          panel.background = element_rect(color = "white"),
          plot.background = element_rect(color = "white"),
          rect = element_rect(size = 0.75, color = "black"), 
          panel.border = element_rect(size = 0.75, color = "black"), 
          panel.grid.major.x = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
          axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
    ylab("") +
    xlab("Read Tags (%)") 
  
  # 3. STAR mapping statistics
  p.star <- select(mydata, NAME, P_uniquely_mapped_reads, P_multiple_Loci, P_too_short_read) %>%
    gather(key, val, -NAME) %>%
    mutate(key = case_when(key == "P_uniquely_mapped_reads" ~ "Uniquely Mapped", 
                           key == "P_multiple_Loci" ~ "Multi-Mapped", 
                           key == "P_too_short_read" ~ "Unmapped"),
           key = factor(as.factor(key), levels = c("Uniquely Mapped", "Multi-Mapped", "Unmapped")),
           FILL = paste(key, NAME)) %>%
    mutate(val = as.numeric(as.character(val))) %>%
    ggplot(aes(x = NAME, y = val, fill = NAME)) +
    geom_violin() +
    geom_boxplot(width = 0.1) +
    ylab("Reads (%)") +
    xlab("") +
    facet_wrap("key", nrow = 1, scales = "free_y") +
    theme_cowplot(font_size = 28) +
    theme(legend.position = "none",
          panel.background = element_rect(color = "white"),
          plot.background = element_rect(color = "white"),
          rect = element_rect(size = 0.75, color = "black"), 
          strip.text = element_text(size = 18, face = "bold"),
          panel.border = element_rect(size = 0.75, color = "black"), 
          axis.text.x = element_text(size = 20, angle = 45, hjust = 1),
          panel.grid.major.y = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
          axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
    scale_fill_manual(values = mycols) 
  
  # 4. Combine and save the graphics
  ggsave(bg = "white",
         plot_grid(
           plot_grid(p.gene, p.star, 
                     nrow = 1, labels = c("a", "b"), label_size = 28, rel_widths = c(0.32, 0.67)), 
           p.resqc, nrow = 2, label_size = 28, labels = c("", "c")) + theme(panel.background = element_rect(color = "white"),
                                                                            plot.background = element_rect(color = "white")), 
         filename = paste0("/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/", title, ".pdf"),
         dpi = 450, width = 32, height = 20)
  
  return(list(p.gene, p.star,p.resqc))
}

```


```{r Load Data, echo = FALSE, include = TRUE, warning = FALSE, comment = FALSE, message = FALSE, cache = FALSE}

# 0. Load the global sample sheet
SAMPLESHEET <- read.delim("/home/vincent.hahaut/Desktop/GITHUB/IOB/sampleSheet/FS_GLOBAL_SAMPLESHEET_17112021.txt", sep = "\t")

# 1. STAR - Statistics (getSTARlog)
STAR.log <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/STAR.logs.2021-12-05.rds.bz")

# 2. MEDIAN GENE COUNTS - FEATURECOUNTS - DOWNSAMPLINGS (sceToExpressedGenes)
med.exon <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/downsampling_medianGenes_featureCounts.exon.all.2021-12-05.rds.bz") %>%
  processExpressGenes(mydata = ., SAMPLESHEET = SAMPLESHEET, STAR.log = STAR.log)

# 3. RAW SCE OBJECTS
# SingleCellObject containing the read counts associated to each cell
sce <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/downsampling_featureCounts.exon.all.2021-12-05.rds.bz")

# 4. Salmon (sceToExpressedGenes)
# Downsampled to 250 or 500K raw reads
sce.sal.500K.med <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/downsampling_medianGenes_salmon.all.500000.2022-03-25.rds.bz")
# sce.sal.250K.med <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/downsampling_medianGenes_salmon.all.250000.2021-12-05.rds.bz")

# 5. RESQC - READ DISTRIBUTION (getRESeQC_readDistribution)
resqc <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/ReSQC.distribution.2021-12-05.rds.bz")

# 6. RESQC - GENE BODY COVERAGES (getRESeQC_coverage)
cover <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/ReSQC.coverage.2021-12-05.rds.bz")

# 7. Estimated FS Timing (see supplementary excel file)
timing <- read_tsv("/home/vincent.hahaut/data_storage/FS_intermediate_file/FS_timing.txt")


```

## GENE EXPRESSIONG HEK - 96w

```{r main gene expression 96, eval = FALSE, echo = F, include = FALSE, warning = FALSE, comment = F, message = F, cache = FALSE, fig.align="center}

# 1. Gather the HEK cells processed in 96w and add group info
# Tidy the information
# Filter out low quality cells
mydata <- med.exon %>%
  filter(PLATE %in% c("2", "3", "5", "6", "10", "12", "247", "309")) %>%
  mutate(NAME = case_when(
    PLATE == "247" & grepl("LV1", ID) | PLATE == "309" & dCTP == "3.3" ~ "FS - 5 µL (Miniaturized)",
    PLATE %in% c("2", "3") ~ "SS2",
    PLATE %in% c("5","6") ~ "SS3",
    PLATE %in% c("10", "12") | PLATE == "247" & grepl("LV5", ID) ~ "FS - 25 µL"),
    DEPTH = factor(DEPTH, levels = 
                     c("5000", "10000","20000","40000","50000","75000",
                       "100000","125000","250000","375000","500000", "625000",
                       "750000","1000000","1250000","1500000")),
    NAME = factor(NAME, levels = c("SS2", "SS3", "FS - 25 µL", "FS - 5 µL (Miniaturized)"))) %>%
  filter(!is.na(NAME), as.numeric(DEPTH) < 1000000, as.numeric(P_too_short_read) < 25)
write.table(mydata$ID, "/home/vincent.hahaut/Desktop/IDs.txt", col.names = F, row.names = F, quote = F, append = TRUE)

# Min diff visible

FS <- filter(mydata, DEPTH == 250000, NAME == "FS - 5 µL (Miniaturized)")
p <- vector()
diff.count <- .96
for(i in 1:100){
  FS.tmp <- sample_n(FS, 16)
  df <- data.frame(
    count_0 = c(FS.tmp$count_0, as.numeric(rnorm(mean = 10118.28*diff.count, sd = 636.6911*diff.count, n = 16))),
    GROUP = c(rep("data", 16), rep("simulated", 16)))
  p <- c(p, rstatix::wilcox_test(df, count_0 ~ GROUP)$p)
}
           

# 2. Gene Counts

# 2.1. Statistical comparison between groups

# 2.1.1. Test normality
shapiro.test(mydata$count_0)

# 2.1.2. Kruskal-wallis / Dunn's test
kruskal.test(data = mydata %>% filter(DEPTH == 500000), count_0 ~ NAME)

mydata.tmp <- mydata %>%
  filter(DEPTH == 500000 & NAME != "FS - 5 µL (Miniaturized)") %>%
  dplyr::select(count_5, count_0, NAME) %>%
  pivot_longer(-NAME, names_to = "threshold", values_to = "genes") %>%
  mutate(group = paste0(threshold, "_", NAME), 
         group = factor(group, levels = c(
           paste0("count_0_", c("SS2", "SS3", "FS - 25 µL")),
           paste0("count_5_", c("SS2", "SS3", "FS - 25 µL")))))


p.signif <- mydata.tmp %>% 
  group_by(threshold) %>%
  rstatix::dunn_test(genes ~ group, p.adjust.method = "bonferroni") %>%
  filter(p.adj < 0.05) %>% 
  rstatix::add_y_position(fun = "max", scales = "free") %>%
  mutate(p.adj.signif = formatC(p.adj, format = "e", digits = 2),
         y.position = rep(c(12000, 13000, 14000), 2))

# 2.2. Display results
p1.500K <- ggplot() +
  geom_hline(yintercept = seq(0,12000,1000), linetype = "dotted", color = "darkgrey") +
  geom_violin(data = mydata.tmp, aes(y = genes, x = group, fill = NAME)) +
  geom_boxplot(data = mydata.tmp, aes(y = genes, x = group, fill = NAME), width = 0.1, position = position_dodge(width = 0.9)) +
  scale_fill_manual(values = c("FS - 25 µL" = "#E41A1C",
                               "SS2" = "#377EB8",
                               "SS3" = "#4DAF4A")) +
  scale_y_continuous(breaks = seq(0,12000,1000)) +
  scale_x_discrete(labels = c(">0 read", ">5 reads")) +
  ylab("Genes Detected") + 
  xlab("") +
  theme_cowplot(font_size = 28) +
  theme(legend.title = element_blank(),
        axis.text = element_text(size = 22), 
        axis.title = element_text(size = 24),
        legend.position = "none",
        axis.text.x = element_blank(), axis.ticks.x = element_blank(),
        legend.text = element_text(size = 22))  +
  scale_y_continuous(breaks = seq(0,12000,1000), labels = c("0", "1,000", "2,000", "3,000", "4,000", "5,000", "6,000", "7,000", "8,000", "9,000", "10,000", "11,000", "12,000")) +
  ggpubr::stat_pvalue_manual(p.signif, label = "p = {p.adj.signif}",  tip.length = 0.01, size = 4) +
  geom_text(data = data.frame(NAME = c("SS2", "SS3", "FS")), aes(y = c(200,200,200), x = c(1,2,3), label = c("n=80", "n=51", "n=105")), size = 5) 

# 2.3. Save results
ggsave(p1.500K, 
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_exonCounts.pdf", bg = "white", dpi = 450, width = 7, height = 6)

# 3. General Statistics
summaryPlots(filter(mydata, DEPTH == 500000), title = "HEK_96w_SUMMARY", mycols = c("FS - 25 µL" = "#E41A1C",
                                                           "SS2" = "#377EB8",
                                                           "SS3" = "#4DAF4A",
                                                           "FS - 5 µL (Miniaturized)" = "#FF850D"))

# 4. Percentage of intronic read tags
p.intron.hek <- right_join(resqc, select(mydata %>% filter(DEPTH == "500000"), ID, NAME)) %>%
  select(ID, UTR3_Exons, UTR5_Exons, CDS_Exons, Introns, TES_down_10kb, TSS_up_10kb, totalReadTags, NAME) %>%
  mutate(Intergenic = totalReadTags - UTR3_Exons - UTR5_Exons - CDS_Exons - Introns - TES_down_10kb - TSS_up_10kb) %>%
  gather(key, val, -ID, -totalReadTags, -NAME) %>%
  mutate(perc = 100*val/totalReadTags,
         NAME = case_when(NAME == "FS - 5 µL (Miniaturized)" ~ "F", 
                          NAME == "SS2" ~ "SS2", 
                          NAME == "SS3" ~ "SS3", 
                          NAME == "FS - 25 µL" ~ "FS - 25 µL"),
         NAME = factor(as.factor(NAME), levels = c("FS - 25 µL", "F", "SS2", "SS3")),
         key = factor(as.factor(key), levels = c("TES_down_10kb", "TSS_up_10kb", "UTR3_Exons", "UTR5_Exons", "Intergenic", "Introns", "CDS_Exons"))) %>%
  # Only display the CDS Exon / intron information
  ggplot(aes(x = NAME, y = perc, fill = NAME)) +
  geom_violin() +
  geom_boxplot(width =  0.1) +
  scale_fill_manual(values = c("FS - 25 µL" = "#E41A1C",
                               "SS2" = "#377EB8",
                               "SS3" = "#4DAF4A",
                               "F" = "#FF850D")) +
  theme_cowplot(font_size = 28) +
  scale_y_continuous(breaks = seq(0,65,10)) +
  facet_wrap("key", scales = "free_x", nrow = 1) +
  theme(legend.position = "none",
        rect = element_rect(size = 0.75, color = "black"), 
        panel.border = element_rect(size = 0.75, color = "black"), 
        panel.grid.major.y = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
        axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
  xlab("") +
  ylab("Read Tags (%)") 

ggsave(p.intron.hek, 
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_introns.pdf", bg = "white", dpi = 450, height = 5, width = 10)

# 5. Gene counts - downsampling

# 5.1. Median and SD per group and depth
mydata.down <- mydata %>%
  mutate(DEPTH = as.numeric(as.character(DEPTH))) %>%
  filter(DEPTH <= 500000) %>%
  dplyr::select(count_0, NAME, DEPTH) %>%
  group_by(NAME, DEPTH) %>%
  summarise(m = median(count_0), sd = sd(count_0)) 

# 5.2. Display results
p2 <- mydata.down %>%
  ggplot(aes(y = m, x = DEPTH, color=NAME, group = NAME)) +
  geom_hline(yintercept = seq(0,12000,1000), linetype = "dashed", color = "darkgrey") +
  geom_line(size = 1.25, position=position_dodge(width=5000)) +
  geom_point(size = 4, position=position_dodge(width=5000), alpha = 0.75)+
  geom_errorbar(aes(ymin=m-sd, ymax=m+sd, color = NAME), position=position_dodge(width=5000), alpha = 0.75) +
  scale_color_manual(values = c("FS - 25 µL" = "#E41A1C",
                                "SS2" = "#377EB8",
                                "FS - 5 µL (Miniaturized)" = "#FF850D",
                                "SS3" = "#4DAF4A",
                                "SSsc" = "#984EA3")) + 
  scale_fill_manual(values = c("FS - 25 µL" = "#E41A1C",
                               "SS2" = "#377EB8",
                               "FS - 5 µL (Miniaturized)" = "#FF850D",
                               "SS3" = "#4DAF4A",
                               "SSsc" = "#984EA3")) + 
  scale_y_continuous(breaks = seq(0,12000,1000), labels = c("0","1,000","2,000","3,000","4,000","5,000","6,000","7,000","8,000","9,000" ,"10,000", "11,000", "12,000")) +
  scale_x_continuous(breaks = c(5000, 10000,20000,40000,50000,75000,100000,125000,250000,375000,500000), labels = c(c("", "", "20K","","50K","75K","100K", "125K","250K","375K","500K"))) +
  ylab("Genes Detected") + 
  xlab("Downsampled Reads") +
  theme_cowplot() +
  theme(legend.title = element_blank(),
        legend.position = "none",
        axis.text.y = element_text(size = 18),
        axis.text.x = element_text(size = 15, angle = 45, hjust = 1),
        axis.title = element_text(size = 24),
        legend.text = element_text(size = 14)) 

ggsave(p2, 
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_exonCounts.downsampling.pdf", bg = "white", dpi = 450, width = 10, height = 6)


# 6. Gene diversity 
# 100-times: take 10 cells, get the genes expressed (>=2 reads) in >=2 cells
# Overlaps between groups, gene types, gene properties (GC, length)
# Plus a few things not used in the final version of the manuscript
set.seed(42)

# 6.1. Load the gene information
# Compute the GC content and actual gene length of the tested genes
# Not directly provided with gencode
if(!file.exists("/home/vincent.hahaut/data_storage/REFERENCES/hsap_fastsmart/REFERENCE/GTF/gencode.v34.exon.GC_lengths.tsv")){
  # GC content Compute
  # 99% from https://github.com/dpryan79/Answers/blob/master/SEQanswers_42420/GTF2LengthGC.R
  library(GenomicRanges)
  library(rtracklayer)
  library(Rsamtools)
  
  GTFfile = "/home/vincent.hahaut/data_storage/REFERENCES/hsap_fastsmart/REFERENCE/GTF/gencode.v34.primary_assembly.annotation.gtf"
  FASTAfile = "/home/vincent.hahaut/data_storage/REFERENCES/hsap_fastsmart/REFERENCE/DNA/GRCh38.primary_assembly.genome.fa"
  
  # 6.1.1. Load the annotation and reduce it
  GTF <- import.gff(GTFfile, format="gtf", genome="GRCh38", feature.type="exon")
  grl <- GenomicRanges::reduce(split(GTF, elementMetadata(GTF)$gene_name))
  reducedGTF <- unlist(grl, use.names=T)
  elementMetadata(reducedGTF)$gene_name <- rep(names(grl), elementNROWS(grl))
  
  # 6.1.2. Open the fasta file
  FASTA <- FaFile(FASTAfile)
  open(FASTA)
  
  # 6.1.3. Add the GC% and gene lengths
  elementMetadata(reducedGTF)$nGCs <- letterFrequency(getSeq(FASTA, reducedGTF), "GC")[,1]
  elementMetadata(reducedGTF)$widths <- width(reducedGTF)
  
  # 6.1.4. Create a list of the ID/GC/length
  calc_GC_length <- function(x) {
    nGCs = sum(elementMetadata(x)$nGCs)
    width = sum(elementMetadata(x)$widths)
    c(width, nGCs/width)
  }
  output <- t(sapply(split(reducedGTF, elementMetadata(reducedGTF)$gene_name), calc_GC_length))
  output <- rownames_to_column(as.data.frame(output), "gene_name")
  colnames(output) <- c("gene_name", "Length", "GC")
  
  # 6.1.5. Save results
  write.table(output, file="/home/vincent.hahaut/data_storage/REFERENCES/hsap_fastsmart/REFERENCE/GTF/gencode.v34.exon.GC_lengths.tsv", quote = FALSE, sep="\t")
} else {
  # 6.1.6. If already created - load it
  gc.perc <- read.table(file="/home/vincent.hahaut/data_storage/REFERENCES/hsap_fastsmart/REFERENCE/GTF/gencode.v34.exon.GC_lengths.tsv", header = T, sep="\t")
}

# 6.2. Resampling

if(!file.exists("/home/vincent.hahaut/Desktop/gene.diversity.rds")){
  
  # 6.2.1. Build a samplesheet 
  # Contains the information related to the cells to resample (unique ID, depth, group)
  sp <- data_frame(
    ID = colnames(sce),
    ID.simplified = str_replace(string = ID, 
                                pattern = paste0(paste0("_", c(5000, 10000, 20000, 40000, 50000, 75000, 100000, 125000, 250000, 375000, 500000), "_1$"), collapse = "|"), ""),
    DEPTH = str_replace_all(pattern = "^_|_1$", "", 
                            string = str_extract(pattern = paste0(paste0("_", c(5000, 10000, 20000, 40000, 50000, 75000, 100000, 125000, 250000, 375000, 500000), "_1$"), collapse = "|"), colnames(sce)))) %>%
    # 6.2.1.1. Filter Cells with at least 500K
    filter(ID.simplified %in% str_replace(colnames(sce[,grepl("_500000_1", colnames(sce))]), "_500000_1", "")) %>%
    # 6.2.1.2. Get HEK cells of interest
    filter(str_replace("X", "", string=str_replace(ID.simplified, "\\.", "-")) %in% mydata$ID) %>%
    # 6.2.1.3. Add Groups
    mutate(GROUP = case_when(
      grepl("X247", ID) & grepl("LV1", ID) | grepl("X309", ID) ~ "FS - 5 µL (Miniaturized)",
      grepl("X2_|X3_", ID)  ~ "SS2",
      grepl("X5_|X6_", ID)  ~ "SS3",
      TRUE ~ "FS - 25 µL"))
  
  # check-ups
  mydata %>% filter(DEPTH == 500000) %>% group_by(NAME) %>% summarise(n())
  sp %>% filter(DEPTH == 500000)  %>% group_by(GROUP) %>% summarise(n())
  sp %>% group_by(DEPTH, GROUP) %>% summarise(n())
  
  # 6.2.2. Resampling
  
  # For each depth, resample 100-times 10 cells
  genes <- list()
  overlaps.res <- genecontent.res <- gc.expressed.genes <- data.frame()
  set.seed(42)
  for(k in  c(5000, 10000, 20000, 40000, 50000, 75000, 100000, 125000, 250000, 375000, 500000)){
    
    for(i in 1:100){
      message(i)
      # 6.2.2.1. Select 10 random cells
      index.tmp <- sp %>% filter(DEPTH == k) %>% group_by(GROUP) %>% sample_n(10) %>% ungroup()
      genes <- list()
      for(j in c("FS - 25 µL", "SS3", "SS2", "FS - 5 µL (Miniaturized)")){
        # 6.2.2.2. Extract them from the single-cell object
        sc <- sce[,index.tmp$ID[index.tmp$GROUP == j]]
        # 6.2.2.3. Get the genes expressed
        genes[[j]] <- row.names(sc)[which(rowSums(counts(sc) > 1) > 1)]
      }
      
      # 6.2.2.4. Overlap Between Methods
      # Only at 500K
      if(k == 500000){
        overlaps <- VennDiagram::calculate.overlap(list(
          `FS - 25 µL` = genes[[1]], 
          SS3 = genes[[2]], 
          SS2 = genes[[3]], 
          `FS - 5 µL (Miniaturized)` = genes[[4]])) 
        
        totals <- sapply(overlaps, length)
        #https://stackoverflow.com/questions/37450698/the-output-order-of-function-calculate-overlap
        overlaps.res <- bind_rows(overlaps.res,
                                  data_frame(
                                    FS_total = length(genes[[1]]),    
                                    SS3_total = length(genes[[2]]),
                                    SS2_total = length(genes[[3]]),
                                    FSmin_total =  length(genes[[4]]),
                                    FS_SS3_SS2_FSmin  = totals["a6"],
                                    FS_SS3_SS2 = totals["a12"],
                                    FS_SS3_FSmin = totals["a11"],
                                    FS_SS2_FSmin = totals["a5"],
                                    SS3_SS2_FSmin = totals["a7"],
                                    FS_SS3 = totals["a15"],
                                    FS_SS2 = totals["a4"],
                                    FS_FSmin = totals["a10"],
                                    SS3_SS2 = totals["a13"],
                                    SS3_FSmin = totals["a8"],
                                    SS2_FSmin = totals["a2"],
                                    FS = totals["a9"],
                                    SS3 = totals["a14"],
                                    SS2 = totals["a1"],
                                    FSmin = totals["a3"],
                                    time = i))
        
      }
      
      # 6.2.2.5. Get types / content
      
      # Local Functions
      
      #' getGeneContent
      #' 
      #' Compute the percentage of each gene type in a series of selected gene names
      #' @param mygenes = vector of gene names
      #' @param gtf = opened GTF file (rtracklayers)
      getGeneContent <- function(mygenes = NULL, gtf = NULL){
        as_tibble(gtf) %>%
          filter(type == "gene") %>%
          dplyr::select(gene_name, gene_type) %>%
          # Get rid of the 78 duplicated gene names in the GTF
          filter(gene_name %in% gene_name[!duplicated(gene_name)]) %>%
          # Select the genes expressed
          filter(gene_name %in% mygenes) %>%
          mutate(total = n()) %>%
          # Compute the percentage of each gene type
          group_by(gene_type) %>%
          summarise(n = n(),
                    perc = 100*n/unique(total)) %>%
          ungroup() %>%
          return()
      }
      
      #' getGCLengthcontent
      #' 
      #' Associate gene names to their respective gene length / GC content
      #' @param mygenes = vector of gene names
      #' @param gc.perc = gc.perc (see above)
      getGCLengthcontent <- function(mygenes = NULL, gc.perc = gc.perc){
        left_join(data_frame(gene_name = mygenes),
                  gc.perc) %>% return()
      }
      
      # Associate genes found in ALL groups or unique to a group
      genecontent.res <- bind_rows(
        genecontent.res,
        getGeneContent(mygenes = genes[[1]], gtf = gtf) %>% mutate(time = i, DEPTH = k, GROUP = "FS - 25 µL", TYPE = "ALL"),
        getGeneContent(mygenes = genes[[2]], gtf = gtf) %>% mutate(time = i, DEPTH = k,GROUP = "SS3", TYPE = "ALL"),
        getGeneContent(mygenes = genes[[3]], gtf = gtf) %>% mutate(time = i, DEPTH = k,GROUP = "SS2", TYPE = "ALL"),
        getGeneContent(mygenes = genes[[4]], gtf = gtf) %>% mutate(time = i, DEPTH = k,GROUP = "FS - 5 µL (Miniaturized)", TYPE = "ALL"),
        getGeneContent(mygenes = overlaps$a9, gtf = gtf) %>% mutate(time = i, DEPTH = k,GROUP = "FS - 25 µL", TYPE = "UNIQUE"),
        getGeneContent(mygenes = overlaps$a14, gtf = gtf) %>% mutate(time = i,DEPTH = k, GROUP = "SS3", TYPE = "UNIQUE"),
        getGeneContent(mygenes = overlaps$a1, gtf = gtf) %>% mutate(time = i, DEPTH = k,GROUP = "SS2", TYPE = "UNIQUE"),
        getGeneContent(mygenes = overlaps$a3, gtf = gtf) %>% mutate(time = i, DEPTH = k,GROUP = "FS - 5 µL (Miniaturized)", TYPE = "UNIQUE"))
      
      # GC content
      gc.expressed.genes <- bind_rows(gc.expressed.genes,
                                      tibble(gene_name = overlaps$a9) %>% left_join(., gc.perc, by = "gene_name") %>% 
                                        summarise(m.gc = mean(GC), m.l = mean(Length), DEPTH = k,time = i, GROUP = "FS - 25 µL"),
                                      tibble(gene_name = overlaps$a14) %>% left_join(., gc.perc, by = "gene_name") %>% 
                                        summarise(m.gc = mean(GC), m.l = mean(Length), DEPTH = k,time = i, GROUP = "SS3"),
                                      tibble(gene_name = overlaps$a1) %>% left_join(., gc.perc, by = "gene_name") %>% 
                                        summarise(m.gc = mean(GC), m.l = mean(Length), DEPTH = k,time = i, GROUP = "SS2"),
                                      tibble(gene_name = overlaps$a3) %>% left_join(., gc.perc, by = "gene_name") %>% 
                                        summarise(m.gc = mean(GC), m.l = mean(Length), DEPTH = k, time = i, GROUP = "FS - 5 µL (Miniaturized)"))
      
    }
    
    
  } 
  
  diversity <- list(overlaps.res = overlaps.res,
                    genecontent.res = genecontent.res,
                    gc.expressed.genes = gc.expressed.genes)
  write_rds(diversity, "/home/vincent.hahaut/Desktop/gene.diversity.rds")
  
} else {
  
  diversity <- read_rds("/home/vincent.hahaut/Desktop/gene.diversity.rds") 
  overlaps.res = diversity[["overlaps.res"]]
  genecontent.res = diversity[["genecontent.res"]]
  gc.expressed.genes = diversity[["gc.expressed.genes"]]
}

# 7. Genes overlapping between methods - Venn Diagramm

# 7.1. Tidy
overlaps.grouped <- overlaps.res %>%
  pivot_longer(names_to = "CATEGORY", values_to = "nGenes", -time) %>%
  group_by(CATEGORY) %>%
  summarise(m = mean(nGenes),
            sd = sd(nGenes)) %>%
  ungroup() %>%
  mutate(lab = paste0(floor(m), "±", floor(sd)))

## 7.2. Manually create the Venn diagramm based on these results
tiff("/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/vennDiagram.pdf", width = 800, height = 800, bg = "white", res = 300)
VennDiagram::draw.quad.venn(
  area1 = 0,
  area2 = 0, 
  area3 = 0,
  area4 = 0,
  n12 = 0,
  n13 = 0,
  n14 = 0, 
  n23 = 0,
  n24 = 0,
  n34 = 0,
  n123 = 0, 
  n124 = 0,
  n134 = 0,
  n234 = 0,
  n1234 = 0,
  category = c("", "", "", ""), lwd = 0.75,
  fill = alpha(c("#E41A1C", "#377EB8", "#4DAF4A", "#FF850D"),0.5),
  cex = 0,
  cat.cex = 0
)
dev.off()

# 8. Global Gene Diversity
# Number of genes >1 read >1 cell in each iteration

# 8.1. Tidy
test <- pivot_longer(overlaps.res, names_to = "CATEGORY", values_to = "nGenes", -time) %>% 
  filter(grepl(CATEGORY, pattern = "total")) %>%
  mutate(GROUP = str_replace(CATEGORY, "_total", ""),
         GROUP = case_when(GROUP == "FS" ~ "FS - 25 µL",
                           GROUP == "FSmin" ~  "FS - 5 µL",
                           GROUP == "SS3" ~ "SS3", 
                           GROUP == "SS2"  ~  "SS2")) 

# 8.2. Statistical test
shapiro.test(test$nGenes)

p.signif <- test %>% 
  rstatix::dunn_test(nGenes ~ GROUP) %>%
  rstatix::add_y_position(fun = "max", scales = "free") %>%
  mutate(p.adj.signif = formatC(p.adj, format = "e", digits = 2))

# 8.3. Visualize
p.diversity <- ggplot() +
  geom_hline(yintercept = seq(11000,13500,1000), linetype = "dashed", color ="darkgrey") +
  geom_violin(data = test, aes(x = GROUP, y = nGenes, fill = GROUP)) +
  geom_boxplot(data = test, aes(x = GROUP, y = nGenes, fill = GROUP), width = 0.1) +
  ggpubr::stat_pvalue_manual(p.signif, label = "p = {p.adj.signif}",  tip.length = 0.01, size = 4) +
  theme_cowplot(font_size = 20) +
  theme(legend.position = "none", axis.text.x = element_text(size = 14)) +
  xlab("") +
  scale_y_continuous(breaks = seq(11000,13500,1000), labels = c("11,000", "12,000", "13,000")) +
  ylab("Gene Diversity") +
  scale_fill_manual(values = c("FS - 25 µL" = "#E41A1C",
                               "SS2" = "#377EB8",
                               "FS - 5 µL" = "#FF850D",
                               "SS3" = "#4DAF4A",
                               "SSsc" = "#984EA3")) 
ggsave(p.diversity, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_diversity.pdf", width = 6, height = 6, dpi= 450, bg = "white")

# 9. Gene types

# 9.1. Tidy
# Get all the genes from each iteration
# DEPTH 500K
# Compute % of each gene type
test <- genecontent.res %>%
  filter(TYPE == "ALL", DEPTH == 500000,
         gene_type %in% c("lncRNA","protein_coding", "processed_pseudogene", "transcribed_unprocessed_pseudogene", "TEC", "unprocessed_pseudogene")) %>%
  mutate(gene_type = case_when(gene_type == "TEC" ~ "To be\nExperimentally\nConfirmed",
                               gene_type == "unprocessed_pseudogene" ~ "Unprocessed\nPseudogenes",
                               gene_type == "lncRNA"  ~ "lncRNA",
                               gene_type == "protein_coding" ~ "Protein Coding",
                               gene_type == "processed_pseudogene" ~ "Processed\nPseudogenes",
                               gene_type == "transcribed_unprocessed_pseudogene" ~ "Transcribed\nUnprocessed\nPseudogenes"),
         gene_type = factor(gene_type, levels = c("Protein Coding", "lncRNA", "Processed\nPseudogenes", "Transcribed\nUnprocessed\nPseudogenes", "To be\nExperimentally\nConfirmed", "Unprocessed\nPseudogenes")))

# 9.2. Statstical significance
# Add the info required to display these results
p.signif <- test %>%
  group_by(gene_type) %>%
  rstatix::dunn_test(n ~ GROUP, p.adjust.method = "bonferroni") %>%
  filter(p.adj < 0.05) %>%
  rstatix::add_xy_position(fun = "max", scales = "free_y") %>%
  mutate(p.adj.signif = formatC(p.adj, format = "e", digits = 2))


# 9.3. Visualize results
p.gene.content <- ggplot() +
  geom_violin(data = test, aes(x = GROUP, y = n, fill = GROUP)) +
  geom_boxplot(data = test, aes(x = GROUP, y = n, fill = GROUP), width = 0.1, position = position_dodge(0.9)) +
  xlab("") +
  theme_bw(base_size = 24) +
  theme(legend.position = "none",
        legend.text = element_text(size = 18),
        legend.title = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  scale_fill_manual(values = c("FS - 25 µL" = "#E41A1C",
                               "SS2" = "#377EB8",
                               "FS - 5 µL (Miniaturized)" = "#FF850D",
                               "SS3" = "#4DAF4A",
                               "SSsc" = "#984EA3")) + 
  scale_y_continuous(label = scales::comma) +
  ylab("Genes Detected") +
  ggpubr::stat_pvalue_manual(p.signif, label = "p = {p.adj.signif}",  tip.length = 0.01, size = 4) +
  facet_wrap("gene_type", ncol = 2, scales = "free") 

ggsave(p.gene.content, filename = paste0("/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_geneContent.pdf"), dpi = 450, height = 16, width = 12)



# 10. Gene body coverage

# 10.1. Read and load the data related to FS / SS2 HEK
paths.ReSQC.cov <- makePath(seqPaths =   "/home/vincent.hahaut/data_storage/210322_NB551561_0044_AHNTL2AFX2/STAR/", suffix = "ReSQC/ID_INTERNAL_geneBody.all.geneBodyCoverage.txt")
cover.SS3 <- getRESeQC_coverage(paths = paths.ReSQC.cov$paths, sample.ids = paths.ReSQC.cov$sample.ids, threads = 10) %>%
  filter(ID %in% mydata$ID[mydata$NAME == "SS3"])

# 10.2. Tidy (mean / sd)
mycover <- cover %>%
  filter(ID %in% mydata$ID[mydata$NAME %in% c("FS - 25 µL", "SS2")]) %>%
  right_join(mydata %>% select(ID, NAME), by = "ID") %>%
  distinct() %>%
  gather(Percentile, reads, -ID, -NAME) %>%
  filter(reads > 0) %>%
  group_by(ID) %>%
  mutate(percentage = 100*reads / sum(reads)) %>%
  group_by(NAME, Percentile) %>%
  summarise(med.c = median(percentage),
            sd.c = sd(percentage)) %>%
  ungroup() 

# 10.3. Visualize results
p.cover <- ggplot(mycover, aes(x = as.numeric(Percentile), color = as.factor(NAME), fill = as.factor(NAME))) +
  geom_ribbon(aes(ymin = med.c - sd.c, ymax = med.c + sd.c), colour = NA) +
  geom_line(aes(y=med.c), size = 1.25) +
  theme_cowplot(font_size = 28) +
  scale_color_manual(values = c("FS - 25 µL" = "#E41A1C",
                                "SS2" = "#377EB8")) + 
  scale_fill_manual(values = c("FS - 25 µL" = adjustcolor("#E41A1C", alpha = 0.25),
                               "SS2" = adjustcolor("#377EB8", alpha = 0.25))) +
  ylab("Coverage (%)") +
  xlab("Gene Body Percentile (5'->3')") +
  theme(legend.title = element_blank(),
        axis.text = element_text(size = 22),
        axis.title = element_text(size = 24),
        legend.position = "none")

ggsave(p.cover, filename = paste0("/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_geneBodyCoverage.pdf"), bg = "white", dpi = 450, height = 6, width = 8)


# 12. Gene diversity - GC content / length

# 12.1. GC Content

# 12.1.1. Statistical Significance
test <- gc.expressed.genes %>% 
  filter(DEPTH == 500000) %>%
  mutate(GROUP = ifelse(GROUP == "FS - 5 µL (Miniaturized)", "F", GROUP),
         GROUP = factor(GROUP, levels = c("FS - 25 µL", "F", "SS2", "SS3")),
         m.gc = 100*m.gc)

p.signif <- rstatix::dunn_test(data = test, m.gc ~ GROUP, p.adjust.method = "bonferroni") %>%
  filter(p.adj < 0.05) %>%
  rstatix::add_xy_position(fun = "max", scales = "free_y") %>%
  mutate(p.adj.signif = formatC(p.adj, format = "e", digits = 2))

# 12.1.2. Visualize
p.GC <- ggplot() + 
  geom_hline(yintercept = seq(45,54,1), linetype = "dashed", color = "darkgrey") +
  scale_y_continuous(breaks = seq(45,54,1)) +
  geom_violin(data = test, aes(x = GROUP, y = m.gc, fill = GROUP)) + 
  geom_boxplot(data = test, aes(x = GROUP, y = m.gc, fill = GROUP), width = 0.1) +
  scale_fill_manual(values = c("FS - 25 µL" = "#E41A1C",
                               "SS2" = "#377EB8",
                               "F" = "#FF850D",
                               "SS3" = "#4DAF4A")) +
  theme_bw(base_size = 24) + 
  theme(legend.position = "none") +
  ylab("Average GC content (%)") +
  xlab("") +
  theme_cowplot(font_size = 24) +
  theme(legend.position = "none") + 
  ggpubr::stat_pvalue_manual(p.signif, label = "p = {p.adj.signif}",  tip.length = 0.01, size = 4) 

# 12.2. Gene length

# 12.2.1. Statistical Significance
test <- gc.expressed.genes %>% 
  filter(DEPTH == 500000) %>%
  mutate(GROUP = ifelse(GROUP == "FS - 5 µL (Miniaturized)", "F", GROUP)) %>%
  mutate(GROUP = factor(GROUP, levels = c("FS - 25 µL", "F", "SS2", "SS3"))) %>%
  filter(m.l < quantile(m.l, .999)) 

p.signif <- rstatix::dunn_test(data = test, m.l ~ GROUP, p.adjust.method = "bonferroni") %>%
  filter(p.adj < 0.05) %>%
  rstatix::add_xy_position(fun = "max", scales = "free_y") %>%
  mutate(p.adj.signif = formatC(p.adj, format = "e", digits = 2))

# 12.2.2. Visualize
p.Length <- ggplot() + 
  geom_hline(yintercept = seq(2500,4500,500), linetype = "dashed", color = "darkgrey") +
  geom_violin(data = test, aes(x = GROUP, y = m.l, fill = GROUP)) + 
  geom_boxplot(data = test, aes(x = GROUP, y = m.l, fill = GROUP), width = 0.1) +
  scale_fill_manual(values = c("FS - 25 µL" = "#E41A1C",
                               "SS2" = "#377EB8",
                               "F" = "#FF850D",
                               "SS3" = "#4DAF4A")) +
  theme_bw(base_size = 24) + 
  theme(legend.position = "none") +
  ylab("Average Gene Length (bp)") +
  xlab("") +
  scale_y_continuous(breaks = seq(2500,4500,500), label = scales::comma) +
  theme_cowplot(font_size = 24) +
  theme(legend.position = "none") + 
  ggpubr::stat_pvalue_manual(p.signif, label = "p = {p.adj.signif}",  tip.length = 0.01, size = 4) 


ggsave(plot_grid(align = "v", p.GC + theme(legend.position = "none"), p.Length  + theme(legend.position = "none"), ncol = 2), filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_DE_explore.pdf", dpi = 450, height = 6, width = 12, bg = "white")

# 13. Which gene types are gained with increasing sequencing depth ? 

# Tidy + Visualize
# Get the average / sd gene types per sequencing depth
p.genetype.gained <- genecontent.res %>%
  filter(TYPE == "ALL") %>%
  filter(GROUP %in% c("SS3", "FS - 25 µL")) %>%
  filter(gene_type %in% c("lncRNA","protein_coding", "processed_pseudogene", "transcribed_unprocessed_pseudogene", "TEC", "unprocessed_pseudogene", "other")) %>%
  mutate(gene_type = case_when(gene_type == "other" ~ "Other", 
                               gene_type == "TEC" ~ "To be\nExperimentally\nConfirmed",
                               gene_type == "unprocessed_pseudogene" ~ "Unprocessed\nPseudogene",
                               gene_type == "lncRNA"  ~ "lncRNA",
                               gene_type == "protein_coding" ~ "Protein Coding",
                               gene_type == "processed_pseudogene" ~ "Processed\nPseudogene",
                               gene_type == "transcribed_unprocessed_pseudogene" ~ "Transcribed\nUnprocessed\nPseudogene"),
         gene_type = factor(gene_type, levels = c("Protein Coding", "lncRNA", "Processed\nPseudogene", "Transcribed\nUnprocessed\nPseudogene", "To be\nExperimentally\nConfirmed", "Unprocessed\nPseudogene", "Other"))) %>%
  group_by(DEPTH, GROUP, gene_type) %>%
  summarise(m = mean(n), sd = sd(n)) %>%
  ggplot(aes(x = as.factor(DEPTH), y = m , fill = gene_type)) + 
  geom_bar(stat = "identity", position = "dodge", color = "black") +
  geom_errorbar(aes(ymin=m-sd, ymax=m+sd), width=.8, position=position_dodge(.9)) +
  theme_bw(base_size = 18) + 
  theme(legend.text = element_text(size = 14),
        axis.text.x = element_text(angle = 45, hjust = 1), 
        legend.title = element_blank()) +
  scale_fill_manual(values = mycolors) +
  scale_y_continuous(breaks = seq(0,13000,1000)) +
  scale_x_discrete(labels = c("5K", "10K", "20K", "40K", "50K", "75K", "100K", "125K", "250K", "350K", "500K" )) +
  xlab("Downsampled Reads") +
  ylab("Average Detected Genes") +
  facet_wrap("GROUP", nrow = 1)
ggsave(p.genetype.gained + theme(legend.position = "none"), filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_miniaturized_gainedgene.pdf", dpi = 300, height = 5, width = 12, bg = "white")

# 14. HEK isoforms (salmon)
# 500K

# 14.1. Tidy
# DEPTH = 500K
# Add groups
# Select TPM >1 or TPM > 5
med.isoforms <- sce.sal.500K.med %>%
  filter(ID %in% paste0(mydata$ID, "_500000_1")) %>%
  tibble() %>%
  mutate(GROUP = case_when(
    grepl("309_dCTP3x|247_HEK_LV1", ID) ~ "FS - 5 µL",
    grepl("^2_SS2|^3_SS2", ID) ~ "SS2",
    grepl("^5_SS3|^6_SS3", ID) ~ "SS3",
    grepl("^10_std|^12_std|^247_HEK_LV5", ID) ~ "FS - 25 µL")) %>%
  filter(!is.na(GROUP)) %>%
  dplyr::select(GROUP, TPM_1, TPM_5) %>%
  pivot_longer(-GROUP) %>%
  mutate(name = ifelse(name == "TPM_1", "TPM > 1", "TPM > 5"))

# 14.2. Compute statistical significance
p.signif <- med.isoforms %>%
  group_by(name) %>%
  rstatix::dunn_test(value ~ GROUP, p.adjust.method = "bonferroni") %>%
  filter(p.adj < 0.05) %>%
  rstatix::add_xy_position(fun = "max", scales = "free_y") %>%
  mutate(p.adj.signif = formatC(p.adj, format = "e", digits = 2),
         p.adj.signif = ifelse(p.adj.signif == "0.00e+00", "p<1e-100", paste0("p=", p.adj.signif)))

# 14.3. Visualize results
p.isoform.HEK <- ggplot() +
  geom_violin(data = med.isoforms, aes(x = GROUP, y = value, fill = GROUP)) +
  geom_boxplot(data = med.isoforms, aes(x = GROUP, y = value, fill = GROUP), width = 0.1) +
  theme_bw(base_size = 24) +
  facet_wrap("name",  nrow = 1) + 
  scale_y_continuous(breaks = seq(15000, 30000, 5000), labels = c("15,000", "20,000", "25,000", "30,000")) +
  theme(legend.position = "none", axis.text.x = element_text(size = 12)) +
  ggpubr::stat_pvalue_manual(filter(p.signif, name == "TPM > 1"), label = "{p.adj.signif}",  step.increase = 0.01, tip.length = 0.01, size = 4) +
  ggpubr::stat_pvalue_manual(filter(p.signif, name == "TPM > 5"),, label = "{p.adj.signif}",  step.increase = 0.03, tip.length = 0.01, size = 4) +
  xlab("") +
  ylab("Transcripts Detected") +
  scale_fill_manual(values = c("FS - 25 µL" = "#E41A1C",
                               "SS2" = "#377EB8",
                               "FS - 5 µL" = "#FF850D",
                               "SS3" = "#4DAF4A"))
ggsave(p.isoform.HEK, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_isoforms.pdf", dpi = 450, width = 10, height = 8, bg = "white")

# 15. Gained genes against reference depth

# 15.1. tidy
saturation <- mydata %>% 
  select(ID, NAME, DEPTH, count_0) %>%
  # 15.1.1. Filter cells with at least 500K reads
  filter(ID %in% ID[DEPTH == 500000]) %>%
  group_by(ID) %>%
  # 15.1.2. Total number of expressed genes per sample
  mutate(max.genes = count_0[DEPTH == 500000]) %>%
  ungroup() %>%
  # 15.1.3. How many sample at this sequencing depth ? 
  group_by(ID, DEPTH) %>%
  mutate(n = n()) %>%
  # 15.1.4. Remove conditions with < 4 cells
  filter(n < 4) %>%
  # 15.1.5. Proportion of genes compared to the 500K reference
  mutate(prop = count_0/max.genes) %>%
  group_by(DEPTH, NAME) %>%
  # 15.1.6. Median / SD
  summarise(m = median(prop),
            sd = sd(prop)) %>%
  # 15.1.7. Visualize
  ggplot(aes(x = as.numeric(as.character(DEPTH)), color = NAME, group = NAME, fill = NAME)) +
  geom_hline(yintercept = seq(0,1.25,0.1), linetype = "dashed", color = "darkgrey") +
  geom_hline(yintercept = 1, linetype = "dashed", color = "red", size = 0.75) +
  geom_line(aes(y = m), size = 1.25, position=position_dodge(width=5000)) +
  geom_point(aes(y = m), size = 4, position=position_dodge(width=5000), alpha = 0.75)+
  geom_errorbar(aes(ymin=m-sd, ymax=m+sd, color = NAME), position=position_dodge(width=5000), alpha = 0.75) +
  scale_color_manual(values = c("FS - 25 µL" = "#E41A1C",
                                "SS2" = "#377EB8",
                                "FS - 5 µL (Miniaturized)" = "#FF850D",
                                "SS3" = "#4DAF4A",
                                "SSsc" = "#984EA3")) + 
  scale_fill_manual(values = c("FS - 25 µL" = "#E41A1C",
                               "SS2" = "#377EB8",
                               "FS - 5 µL (Miniaturized)" = "#FF850D",
                               "SS3" = "#4DAF4A",
                               "SSsc" = "#984EA3")) +
  scale_x_continuous(breaks = c(5000, 10000,20000,40000,50000,75000,100000,125000,250000,375000,500000,625000,750000, 1000000, 1250000, 1500000), labels = c(c("", "", "","","50K","","100K", "","250K","375K","500K", "625K","750K","1000K","1250K", "1500K"))) +
  scale_y_continuous(breaks = seq(0,1.25,0.1)) +
  ylab("Proportion of\nGenes Detected") + 
  xlab("Downsampled Reads") +
  theme_cowplot() +
  theme(legend.position = "none",
        axis.text.y = element_text(size = 18),
        axis.text.x = element_text(size = 15, angle = 45, hjust = 1),
        axis.title = element_text(size = 24),
        legend.text = element_text(size = 14)) 

ggsave(saturation, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_saturation.pdf", dpi = 450, height = 6, width = 16, bg = "white")

ggsave(plot_grid(p2, saturation, nrow = 2), filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_saturation_downsampling.pdf", dpi = 500, height = 10, width = 11, bg = "white")

```

## Miniaturization

```{r HEK mini, echo = FALSE, include = TRUE, warning = FALSE, comment = FALSE, message = FALSE, cache = FALSE}

mydata <- med.exon %>%
  filter(PLATE == "247") %>%
  mutate(NAME = case_when(
    LYSIS_VOLUME == 0.5 ~ "FS - 2.5 µL",
    grepl("LV1", ID)  ~ "FS - 5 µL",
    grepl("LV5", ID)  ~ "FS - 25 µL"),
    NAME = factor(NAME, levels = c("FS - 2.5 µL", "FS - 5 µL", "FS - 25 µL")),
    sce.ids = paste0(ID, "_500000_1")) %>%
  filter(DEPTH == 500000, as.numeric(P_too_short_read) < 25)

# 2. Gene Counts

# 2.1. Statistical comparison between groups

kruskal.test(data = mydata, count_0 ~ NAME)

# 2.2. Display results
p1.500K <- ggplot() +
  geom_hline(yintercept = seq(0,12000,1000), linetype = "dotted", color = "darkgrey") +
  geom_violin(data = mydata, aes(y = count_0, x = NAME, fill = NAME)) +
  geom_boxplot(data = mydata, aes(y = count_0, x = NAME, fill = NAME), width = 0.1, position = position_dodge(width = 0.9)) +
  scale_fill_manual(values = c("FS - 25 µL" = "#E41A1C",
                               "FS - 5 µL" = "#FF850D",
                               "FS - 2.5 µL" = "#fcdb03")) +
  scale_y_continuous(breaks = seq(0,12000,1000), labels = c("0","1,000","2,000","3,000","4,000","5,000","6,000","7,000","8,000","9,000" ,"10,000", "11,000", "12,000")) +
  ylab("Genes Detected") + 
  xlab("") +
  theme_cowplot(font_size = 28) +
  theme(legend.title = element_blank(),
        axis.text.y = element_text(size = 22), 
        axis.text.x = element_text(size = 18),
        axis.title = element_text(size = 24),
        legend.position = "none", axis.ticks.x = element_blank(),
        legend.text = element_text(size = 22))  +
  geom_text(data = data.frame(NAME = c("SS2", "SS3", "FS")), aes(y = c(200,200,200), x = c(1,2,3), label = c("n=27", "n=20", "n=23")), size = 5) 

# 2.3. Save results
ggsave(p1.500K, 
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_exonCounts_mini.pdf", bg = "white", dpi = 450, width = 7, height = 6)
```

## cDNA yields

```{r HEK yield, echo = FALSE, include = TRUE, warning = FALSE, comment = FALSE, message = FALSE, cache = FALSE}

# 1. Load the cDNA yield information from paltes 2,3,5,6,10,12
# Modify the names for plotting reasons

yield <- read_tsv("/home/vincent.hahaut/data_storage/transfer/Mendeley/cDNA Yields/FS_HEK_cDNAYield_19022021.txt") %>%
  mutate(
    fill = case_when(
      Method == "FS" & `Plate ID` == 12 ~ "FS-19c",
      Method == "FS" & `Plate ID` == 10 ~ "FS-20c",
      Method == "SSII" ~ "SS2",
      Method == "SSIII" ~ "SS3"),
    fill = factor(as.factor(fill), levels = c("FS-20c", "FS-19c", "SS2", "SS3")),
    Method = case_when(
      Method == "FS" ~ "FS - 25 µL\n20c - 19c",
      Method == "SSII" ~ "SS2\n20c",
      Method == "SSIII" ~ "SS3\n20c")) 
# MEAN YIELDS: yield %>% group_by(fill) %>% summarise(mean(`cDNA Yield`, na.rm = T))

pA <- ggplot() +
  geom_hline(yintercept = seq(5,15,5), linetype = "dotted", color = "darkgrey") + 
  geom_boxplot(data = yield, aes(x = Method, y = `cDNA Yield`, fill = yield$fill)) +
  theme_cowplot(font_size = 22) +
  theme(legend.position = "none", axis.text = element_text(size = 24), axis.title.y = element_text(size = 24)) +
  xlab("") +
  ylab("cDNA Yield (ng/µL)") +
  scale_fill_manual(values = c("FS-19c" = colorspace::lighten("#E41A1C", 0.05),
                               "FS-20c" = colorspace::darken("#E41A1C", 0.05),
                               "SS2" = "#377EB8",
                               "SS3" = "#4DAF4A")) +
  geom_text(aes(x = 0.75, label="n = 45"),  y = 0, stat='count', size=5) +
  geom_text(aes(x = 1.25, label="n = 46"),  y = 0, stat='count', size=5) +
  geom_text(aes(x = 2, label="n = 90"),  y = 0, stat='count', size=5) +
  geom_text(aes(x = 3, label="n = 94"),  y = 0, stat='count', size=5) 

ggsave(pA, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_cDNA_yield.pdf", bg = "white", dpi = 450, height = 8, width = 6)

```



## Cell-to-Cell Correlation

```{r main correlation, eval = FALSE, echo = F, include = FALSE, warning = FALSE, comment = F, message = F, cache = FALSE, fig.align="center}

# 1. Select the data of interest
mydata <- med.exon %>%
  filter(PLATE %in% c("2", "3", "5", "6", "10", "12", "247")) %>%
  mutate(NAME = case_when(
    PLATE %in% c("2", "3") ~ "SS2",
    PLATE %in% c("5","6") ~ "SS3",
    PLATE %in% c("10", "12") | PLATE == "247" & grepl("LV5", ID) ~ "FS")) %>%
  filter(!is.na(NAME), DEPTH == 500000, as.numeric(DEPTH) < 1000000, as.numeric(P_too_short_read) < 25) %>%
  mutate(NAME = factor(NAME, levels = c("SS2", "SS3", "FS")),
  # Naming differences between SingleCellObjet and this matrix
         sce.ids = str_replace(paste0("X", ID, "_500000_1"), "-", "."))
write.table(mydata$ID, "/home/vincent.hahaut/Desktop/IDs.txt", col.names = F, row.names = F, quote = F, append = TRUE)

# 2. Select the cells of interest
expressedSet <- sce[,colnames(sce) %in% mydata$sce.ids]

# 3. Add the GROUP info (SS2, SS3 or FS)
expressedSet$GROUP <- mydata$NAME[match(colnames(expressedSet), mydata$sce.ids)]

# 4. For each group, get the expressed genes
# Used a gene set common to all methods to allow comparison between the tau
expressedSet.genes <- sapply(c("SS2", "SS3", "FS"), function(x) 
  row.names(expressedSet)[rowSums(counts(expressedSet)[,which(expressedSet$GROUP == x)]) > 0])

expressedSet.genes <- intersect(intersect(expressedSet.genes$SS2, expressedSet.genes$SS3), expressedSet.genes$FS)
sce.expressed <- expressedSet[row.names(expressedSet) %in% expressedSet.genes,]

# Save
newNames <- read_tsv("/home/vincent.hahaut/Desktop/fastqTransform.txt")
newNames$OLD_FASTQ_ID <- str_replace_all(paste0("X", newNames$OLD_FASTQ_ID, "_500000_1"), "-", ".")

colnames(expressedSet) <- newNames$FINAL_FASTQ_ID[match(colnames(expressedSet), newNames$OLD_FASTQ_ID)]

write_rds(expressedSet, "/home/vincent.hahaut/data_storage/transfer/Mendeley/HEK_SS2_SS3_FS_500K.rds")

# 5. Correlations
corrs <- data_frame()
for(j in unique(c("SS2", "SS3", "FS"))){
  
    # 5.1. Subset the singleCellObject for the group of interest
    sce.tmp <- counts(sce.expressed[,sce.expressed$GROUP == j])
    
    # 5.2. Kendall's tau
    p <- pcaPP::cor.fk(sce.tmp) 
    
    # 5.3. Tidy
    corrs <- bind_rows(corrs,
                       data_frame(GROUP = j,
                                  tau = p[upper.tri(p)]))
}


# 6. Compare tau (pairwise)
shapiro.test(sample(corrs$tau, 4000, replace = FALSE))
kruskal.test(data = corrs, tau ~ GROUP)

# Add the info required to display these results
p.signif <- corrs %>%
  rstatix::dunn_test(tau ~ GROUP, p.adjust.method = "bonferroni") %>%
  filter(p.adj < 0.05) %>%
  rstatix::add_xy_position(fun = "max", scales = "free_y") %>%
  mutate(p.adj.signif = formatC(p.adj, format = "e", digits = 2),
         p.adj.signif = ifelse(p.adj.signif == "0.00e+00", "p<1e-100", paste0("p=", p.adj.signif)))

# 7. Visualize results
p6 <- corrs %>%
  mutate(GROUP = ifelse(GROUP == "FS", "FS - 25µL", GROUP)) %>%
  ggplot(aes(x = reorder(GROUP, tau, median), y = tau)) + 
  geom_hline(yintercept = seq(0.40, 0.75, 0.05), linetype = "dotted", color = "darkgrey") +
  geom_violin(aes(fill = GROUP)) +
  geom_boxplot(aes(fill = GROUP), width = 0.1) +
  theme_cowplot(font_size = 18) +
  scale_fill_manual(values = c("FS - 25µL" = "#E41A1C",
                                                         "SS2" = "#377EB8",
                                                         "SS3" = "#4DAF4A",
                                                         "SSsc" = "#984EA3")) +
  theme(legend.position = "none") +
  xlab("") +
  ylab("Cell-to-Cell Correlation (tau)") +
  ggpubr::stat_pvalue_manual(p.signif, label = "{p.adj.signif}",  tip.length = 0.01, size = 4)


ggsave(p6, 
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_kendallCorr.pdf", bg = "white", dpi = 450, width = 5, height = 6)
```

# Miniaturize SS3 / SSsc 

```{r SS3 SSsc miniaturization PBMC, echo = FALSE, include = TRUE, warning = FALSE, comment = FALSE, message = FALSE, cache = FALSE}

# 1. Get the data + Tidy
mydata <- med.exon %>%
  mutate(NAME = case_when(
    PLATE == "32" ~ "SS3 - 10µL",
    PLATE == "139" ~ "SS3 - 5µL",
    PLATE == "39" ~ "SSsc - 50µL",
    PLATE == "160"  ~ "SSsc - 5µL"))  %>%
  filter(!is.na(NAME), DEPTH == 500000) %>%
  mutate(NAME = factor(as.factor(NAME), levels = c("SS3 - 10µL", "SS3 - 5µL", "SSsc - 50µL", "SSsc - 5µL")),
         METHOD = ifelse(METHOD == "SMARTSEQIII", "SS3", "SSsc"))
write.table(mydata$ID, "/home/vincent.hahaut/Desktop/IDs.txt", col.names = F, row.names = F, quote = F, append = TRUE)

# 2. Visualize SS3
p1.ss3 <- ggplot(mydata %>% filter(METHOD == "SS3"), aes(y = count_0, x = NAME, fill = NAME)) +
  geom_hline(yintercept = seq(0,max(mydata$count_0)+500,1000), linetype = "dotted", color = "darkgrey") +
  geom_violin() +
  geom_boxplot(width = 0.1, position = position_dodge(0.9)) +
  scale_fill_manual(values = 
                      c("SS3 - 10µL" = "#4DAF4A", "SS3 - 5µL" = "#afad4a",
                        "SSsc - 50µL" = "#984EA3", "SSsc - 5µL" = "#cfa3d5")) +
  scale_y_continuous(breaks = seq(0,max(mydata$count_0)+500,1000)) +
  scale_y_continuous(limits=c(0,5000)) +
  ylab("Detected Genes") + 
  xlab("") +
  geom_text(aes(x = NAME, label=paste0("n = ", ..count..)), y = -100, stat='count', size=4) +
  theme_cowplot() +
  theme(axis.line=element_line(), 
        legend.position = "none", 
        legend.title = element_blank(),
        strip.text = element_text(size = 20), 
        axis.text = element_text(size = 24), 
        axis.title = element_text(size = 28),
        legend.text = element_text(size = 24),
        rect = element_rect(size = 0.75, color = "black"), 
        panel.border = element_rect(size = 0.75, color = "black"), 
        axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) 

ggsave(p1.ss3, 
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/PBMC_reactionVolume_SS3.pdf", dpi = 300, width = 6, height = 8, bg = "white")

# 3. Visualize Takara
p1.sscr <- ggplot(mydata %>% filter(METHOD == "SSsc"), aes(y = count_0, x = NAME, fill = NAME)) +
  geom_hline(yintercept = seq(0,max(mydata$count_0)+500,1000), linetype = "dotted", color = "darkgrey") +
  geom_violin() +
  geom_boxplot(width = 0.1, position = position_dodge(0.9)) +
  scale_fill_manual(values = 
                      c("SS3 - 10µL" = "#4DAF4A", "SS3 - 5µL" = "#afad4a",
                        "SSsc - 50µL" = "#984EA3", "SSsc - 5µL" = "#cfa3d5")) +
  scale_y_continuous(breaks = seq(0,max(mydata$count_0)+500,1000)) +
  scale_y_continuous(limits=c(0,5000)) +
  ylab("Detected Genes") + 
  xlab("") +
  geom_text(aes(x = NAME, label=paste0("n = ", ..count..)), y = -100, stat='count', size=4) +
  theme_cowplot() +
  theme(axis.line=element_line(), 
        legend.position = "none", 
        legend.title = element_blank(),
        strip.text = element_text(size = 20), 
        axis.text = element_text(size = 24), 
        axis.title = element_text(size = 28),
        legend.text = element_text(size = 24),
        rect = element_rect(size = 0.75, color = "black"), 
        panel.border = element_rect(size = 0.75, color = "black"), 
        axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) 

ggsave(p1.sscr, 
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/PBMC_reactionVolume_SSsc.pdf", dpi = 300, width = 6, height = 8, bg = "white")

```

# LOW AMPLIFICATION

```{r Low Amplification, echo = FALSE, include = TRUE, warning = FALSE, comment = FALSE, message = FALSE, cache = FALSE}

# 1. Select the cells of interest
mydata <- med.exon %>%
  filter(PLATE %in% c("309", "247", "230", "260", "261", "262", "258")) %>%
  # known outliers
  filter(!grepl("262_HEK_LA_12c_01_B3|247_HEK_LV1_D7|247_HEK_LV1_B6", ID)) %>% 
  # Select the representative conditions
  # as demonstrated the amount of Tn5 has little impact on the results
  mutate(NAME = case_when(
    PLATE %in% c("247") & grepl("LV1", ID) | PLATE == "309" & dCTP == "3.3" ~ "FS-19c",
    # PLATE == "230" & CYCLES == "8" & grepl(ID, pattern = "00625") ~ "FS-8c",
    # PLATE == "230" & CYCLES == "8" & grepl(ID, pattern = "0125") ~ "8c_0.125",
    PLATE == "230" & CYCLES == "8" & grepl(ID, pattern = "0015625") ~ "FS-LA-8c",
    PLATE == "260" & CYCLES == "6" & grepl(ID, pattern = "0015625") ~ "FS-LA-6c",
    # PLATE == "260" & CYCLES == "6" & grepl(ID, pattern = "00078125") ~ "FS-6c\nTn5:15.625nL",
    # PLATE == "258" & CYCLES == "4" & grepl(ID, pattern = "00039063") ~ "4c_0.0039063",
    PLATE == "258" & CYCLES == "4" & grepl(ID, pattern = "0015625") ~ "FS-LA-4c",
    # PLATE == "258" & CYCLES == "4" & grepl(ID, pattern = "0078125") ~ "4c_0.078125",
    # PLATE == "261" & CYCLES == "10" & grepl(ID, pattern = "0015625") ~ "10c_0.015625",
    PLATE == "261" & CYCLES == "10" & grepl(ID, pattern = "00625") ~ "FS-LA-10c",
    PLATE == "262" & CYCLES == "12" & grepl(ID, pattern = "01") ~ "FS-LA-12c"
    # PLATE == "262" & CYCLES == "12" & grepl(ID, pattern = "003125") ~ "12c_0.03125"
  )) %>%
  filter(!is.na(NAME), DEPTH == 250000) %>%
  
  mutate(NAME = factor(as.factor(NAME), levels = c("FS-19c", "FS-LA-12c", "FS-LA-10c", "FS-LA-8c", "FS-LA-6c", "FS-LA-4c")))
write.table(mydata$ID, "/home/vincent.hahaut/Desktop/IDs.txt", col.names = F, row.names = F, quote = F, append = TRUE)


# 2. Gene Counts

## COUNT > 0 

# 2.1. Statistical Comparison
mydata.tmp <- mutate(mydata, NAME = reorder(NAME, count_0, median)) %>%
  select(NAME, count_0, count_5) %>%
  pivot_longer(-NAME, names_to = "threshold", values_to = "genes") %>%
  mutate(NAME = factor(NAME, levels = c("FS-19c", "FS-LA-12c", "FS-LA-10c", "FS-LA-8c", "FS-LA-6c", "FS-LA-4c")),
         threshold = ifelse(threshold == "count_0", "count > 0", "count > 5"))

du <-  mydata.tmp %>%
  group_by(threshold) %>%
  rstatix::dunn_test(genes  ~ NAME , p.adjust.method = "bonferroni") %>%
  filter(p.adj < 0.05) %>%
  ungroup() %>%
  rstatix::add_xy_position(fun = "max", scales = "free_y") %>%
  mutate(p.adj.signif = formatC(p.adj, format = "e", digits = 2),
         p.adj.signif = ifelse(p.adj.signif == "0.00e+00", "p<1e-100", paste0("p=", p.adj.signif)),
         y.position = c(seq(12100, 12700,100), seq(12100, 13000,100)))

# 2.2. Visualize
p1 <- ggplot() +
  geom_hline(yintercept = seq(1000,12000,1000), linetype = "dotted", color = "darkgrey") +
  geom_violin(data = mydata.tmp, aes(y = genes, x = NAME, fill = NAME)) +
  geom_boxplot(data = mydata.tmp, aes(y = genes, x = NAME, fill = NAME), width = 0.1) +
  scale_fill_manual(values = c("FS-LA-4c" = "#FED976", 
                               "FS-LA-8c" = "#FD8D3C", 
                               "FS-LA-6c" = "#FEB24C", 
                               "FS-LA-10c" = "#FC4E2A", 
                               "FS-LA-12c" = "#E31A1C", 
                               "FS-19c" = "#B10026")) +
  ylim(c(500,17000)) +
  scale_y_continuous(breaks = seq(1000,12000,2000), labels = c("1,000", "3,000", "5,000", "7,000", "9,000", "11,000")) +
  ylab("Genes Detected") + 
  xlab("") +
  theme_cowplot() +
  theme(legend.position = "none",
        strip.background = element_blank(),
        strip.text = element_text(face = "bold", size = 22, hjust = 0),
        legend.title = element_blank(),
        axis.text.x = element_text(size = 18, angle = 45, hjust = 1), 
        axis.text.y = element_text(size = 18), 
        axis.title = element_text(size = 20)) +
  facet_wrap("threshold", scales = "free_y") +
  geom_text(data = mydata.tmp, aes(x = NAME, label=paste0("n=", ..count..)), y = 200, stat='count', size=5) +
  ggpubr::stat_pvalue_manual(filter(du, threshold == "count > 0"), label = "{p.adj.signif}",  step.increase = 0.05, tip.length = 0.01, size = 3) +  
  ggpubr::stat_pvalue_manual(filter(du, threshold == "count > 5"), label = "{p.adj.signif}",  step.increase = 0.05, tip.length = 0.01, size = 3) 

ggsave(p1, 
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_lowAmplificationCycles.reviewer.pdf", bg = "white", dpi = 450, width = 12, height = 6)

# 3. Gene-body Coverage

# 3.1. Tidy and compute the mean / SD percentage of read accross the gene-body
mycover <- right_join(cover, mydata %>% select(ID, NAME), by = "ID") %>%
  gather(Percentile, reads, -ID, -NAME) %>%
  filter(reads > 0) %>%
  group_by(ID) %>%
  mutate(percentage = 100*reads / sum(reads)) %>%
  group_by(NAME, Percentile) %>%
  summarise(mean.c = mean(percentage),
            sd.c = sd(percentage)) %>%
  ungroup() %>%
  mutate(NAME = str_replace(pattern = "\\n", replacement = " + ",string =  NAME))

# 3.2. Visualize
p.cover <- ggplot(mycover, aes(x = as.numeric(Percentile), color = as.factor(NAME), fill = as.factor(NAME))) +
  geom_ribbon(aes(ymin = mean.c - sd.c, ymax = mean.c + sd.c), colour = NA) +
  geom_line(aes(y=mean.c), size = 1.25) +
  theme_cowplot(font_size = 24) +
  scale_color_manual(values = c("FS-LA-4c" = "#FED976", 
                                "FS-LA-8c" = "#FD8D3C", 
                                "FS-LA-6c" = "#FEB24C", 
                                "FS-LA-10c" = "#FC4E2A", 
                                "FS-LA-12c" = "#E31A1C", 
                                "FS-19c" = "#B10026")) +
  scale_fill_manual(values = c("FS-LA-4c" = alpha("#FED976", 0.15), 
                               "FS-LA-8c" = alpha("#FEB24C", 0.15), 
                               "FS-LA-6c" = alpha("#FD8D3C", 0.15), 
                               "FS-LA-10c" = alpha("#FC4E2A", 0.15),  
                               "FS-LA-12c" = alpha("#E31A1C", 0.15),  
                               "FS-19c" = alpha("#B10026", 0.15))) +
  xlab("Gene Body Percentile (5'->3')") +
  ylab("Percentage Coverage") +
  theme(legend.title = element_blank(),
        axis.text = element_text(size = 22),
        axis.title = element_text(size = 24)) + 
  theme(strip.background = element_blank(),
        strip.text = element_text(face = "bold"))

ggsave(p.cover, filename = paste0("/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_LOWAMP_COVERAGE.pdf"), dpi = 450, height = 8, bg = "white", width = 12)


# 4. Gene Distribution (read tags)

# 4.1. Boxplots
p2 <- filter(resqc, ID %in% mydata$ID) %>%
  select(ID, UTR3_Exons, UTR5_Exons, CDS_Exons, Introns, TES_down_10kb, TSS_up_10kb, totalReadTags) %>%
  mutate(Intergenic = totalReadTags - UTR3_Exons - UTR5_Exons - CDS_Exons - Introns - TES_down_10kb - TSS_up_10kb) %>%
  gather(key, val, -ID, -totalReadTags) %>%
  mutate(perc = 100*val/totalReadTags,
         GROUP = case_when(
           grepl("247_HEK_LV1|^309_", ID) ~ "FS-19c",
           grepl(ID, pattern = "8c_0015625") ~ "FS-LA-8c",
           grepl(ID, pattern = "6c_0015625") ~ "FS-LA-6c",
           grepl(ID, pattern = "4c_0015625") ~ "FS-LA-4c",
           grepl(ID, pattern = "10c_00625") ~ "FS-LA-10c",
           grepl(ID, pattern = "12c_01") ~ "FS-LA-12c")) %>%
  mutate(GROUP = factor(as.factor(GROUP), levels = c("FS-19c", "FS-LA-12c", "FS-LA-10c", "FS-LA-8c", "FS-LA-6c", "FS-LA-4c")),
         key = factor(as.factor(key), levels = c("TES_down_10kb", "TSS_up_10kb", "UTR3_Exons", "UTR5_Exons", "Intergenic", "Introns", "CDS_Exons"))) %>%
  # Bring little info to the picture
  filter(!key %in% c("TES_down_10kb", "TSS_up_10kb")) %>%
  ggplot(aes(y = GROUP, x = perc, fill = GROUP)) +
  geom_boxplot() +
  scale_fill_manual(values = c("FS-LA-4c" = "#FED976", 
                               "FS-LA-6c" = "#FEB24C", 
                               "FS-LA-8c" = "#FD8D3C", 
                               "FS-LA-10c" = "#FC4E2A", 
                               "FS-LA-12c" = "#E31A1C", 
                               "FS-19c" = "#B10026")) +
  theme_cowplot(font_size = 18) +
  facet_wrap("key", scales = "free_x", nrow = 1) +
  theme(legend.position = "none",
        rect = element_rect(size = 0.75, color = "black"), 
        panel.border = element_rect(size = 0.75, color = "black"), 
        panel.grid.major.x = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
        axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
  ylab("") +
  xlab("Read Tags (%)") 

ggsave(p2, bg = "white",
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_lowAmplificationCycles_RESQC.pdf", dpi = 450, width = 12, height = 6)

# 4.2. Lines / Dots per PCR cycles
resqc.tmp <- filter(resqc, ID %in% mydata$ID) %>%
  select(ID, UTR3_Exons, UTR5_Exons, CDS_Exons, Introns, TES_down_10kb, TSS_up_10kb, totalReadTags) %>%
  mutate(Intergenic = totalReadTags - UTR3_Exons - UTR5_Exons - CDS_Exons - Introns - TES_down_10kb - TSS_up_10kb) %>%
  select(ID, Introns, CDS_Exons, Intergenic, totalReadTags) %>%
  gather(key, val, -ID, -totalReadTags) %>%
  mutate(perc = 100*val/totalReadTags,
         GROUP = case_when(
           grepl("247_HEK_LV1|309_dCTP3x", ID) ~ "FS-19c",
           grepl(ID, pattern = "8c_0015625") ~ "FS-LA-8c",
           grepl(ID, pattern = "6c_0015625") ~ "FS-LA-6c",
           grepl(ID, pattern = "4c_0015625") ~ "FS-LA-4c",
           grepl(ID, pattern = "10c_00625") ~ "FS-LA-10c",
           grepl(ID, pattern = "12c_01") ~ "FS-LA-12c")) %>%
  mutate(GROUP = factor(as.factor(GROUP), levels = c("FS-19c", "FS-LA-12c", "FS-LA-10c", "FS-LA-8c", "FS-LA-6c", "FS-LA-4c")),
         key = factor(as.factor(key), levels = c("CDS_Exons", "Intergenic", "Introns"))) 

p2 <-     ggplot(data = resqc.tmp, aes(x = GROUP, y = perc, fill = key)) +
  geom_hline(yintercept = seq(0,65,5), linetype = "dotted", color = "darkgrey")+
  stat_summary(
    fun.y = median,
    geom = 'line',
    size = 1.5,
    aes(group = key, colour = key),
    position = position_dodge(width = 0.9) #this has to be added
  ) +
  geom_boxplot(position = position_dodge(width = 0.9), width = 0.5) +
  scale_y_continuous(breaks = seq(0,70, 10)) +
  theme_cowplot(font_size = 24) +
  theme(legend.title = element_blank(),
        axis.text.x = element_text(size = 24, angle = 45, hjust = 1),
        legend.position = "none") +
  scale_color_brewer(palette = "Dark2") +
  scale_fill_brewer(palette = "Dark2") +
  ylab("Read Tags (%)") +
  xlab("") +
  ggtitle("Read Distribution")


ggsave(p2, 
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_lowAmplificationCycles8CYCLES_RESQC.pdf", dpi = 450, bg = "white", width = 8, height = 6)


# 5. STAR Mapping  Statistics
p3 <- mydata %>%
  select(NAME, P_uniquely_mapped_reads, P_multiple_Loci, P_too_short_read) %>%
  gather(key, val, -NAME) %>%
  mutate(NAME = factor(as.factor(NAME), levels = rev(c("FS-LA-4c", "FS-LA-6c", "FS-LA-8c", "FS-LA-10c", "FS-LA-12c", "FS-19c"))),
         key = case_when(key == "P_uniquely_mapped_reads" ~ "Uniquely Mapped", 
                         key == "P_multiple_Loci" ~ "Multi-Mapped", 
                         key == "P_too_short_read" ~ "Unmapped"),
         key = factor(as.factor(key), levels = c("Uniquely Mapped", "Multi-Mapped", "Unmapped")),
         FILL = paste(key, NAME)) %>%
  mutate(val = as.numeric(as.character(val))) %>%
  ggplot(aes(x = NAME, y = val, fill = key)) +
  geom_hline(yintercept = seq(0,80, 10), linetype = "dotted", color = "darkgrey")+
  stat_summary(
    fun.y = median,
    geom = 'line',
    size = 1.5,
    aes(group = key, colour = key),
    position = position_dodge(width = 0.9) #this has to be added
  ) +
  geom_boxplot(position = position_dodge(width = 0.9), width = 0.5) +
  scale_y_continuous(breaks = seq(0,80, 10)) +
  theme_cowplot(font_size = 24) +
  theme(legend.title = element_blank(),
        axis.text.x = element_text(size = 24, angle = 45, hjust = 1),
        legend.position = "none") +
  scale_color_manual(values = c("#E7298A", "#66A61E", "#E6AB02")) +
  scale_fill_manual(values = c("#E7298A", "#66A61E", "#E6AB02")) +
  ylab("Reads (%)") +
  xlab("") +
  ggtitle("Mapping Statistics")


ggsave(p3, 
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_lowAmplificationCycles_STAR.pdf", bg = "white", dpi = 450, width =  8, height = 6)

ggsave(plot_grid(p2, p3, nrow = 2), bg = "white",
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_lowAmplificationCycles_STAR_COMBINED.pdf", dpi = 450, width = 8, height = 10)

# 6. Gene type content

# 6.1. Load GTF information
genetype <- rtracklayer::import(gtf.path) %>%
  as_tibble() %>%
  select(gene_name, gene_type) %>%
  distinct()

# 6.2. Select of interest cells from count matrix
mysce <- sce[,colnames(sce) %in% mydata$oldID]

# 6.3. Tidy and add the gene type
p.LA.content <- counts(mysce) %>%
  # 6.3.1. Tidy
  rownames_to_column("gene_name") %>%
  pivot_longer(-gene_name) %>%
  left_join(dplyr::select(mydata, oldID, NAME), by = c("name" = "oldID")) %>%
  # 6.3.2. select gene seen in at least 3 cells with 1 or more read
  group_by(gene_name, NAME) %>%
  summarise(n = sum(value >= 1)) %>%
  ungroup() %>%
  filter(n  > 2) %>%
  # 6.3.3. Add genes
  left_join(genetype, by = "gene_name") %>%
  # 6.3.4. Select specific gene types
  filter(gene_type %in% c("lncRNA","protein_coding", "processed_pseudogene", "transcribed_unprocessed_pseudogene", "TEC", "unprocessed_pseudogene")) %>%
  mutate(gene_type = case_when(gene_type == "TEC" ~ "TEC",
                               gene_type == "unprocessed_pseudogene" ~ "Unprocessed\nPseudogene",
                               gene_type == "lncRNA"  ~ "lncRNA",
                               gene_type == "protein_coding" ~ "Protein Coding",
                               gene_type == "processed_pseudogene" ~ "Processed\nPseudogene",
                               gene_type == "transcribed_unprocessed_pseudogene" ~ "Transcribed\nUnprocessed\nPseudogene"),
         gene_type = factor(gene_type, levels = c("Protein Coding", "lncRNA", "Processed\nPseudogene", "Transcribed\nUnprocessed\nPseudogene", "TEC", "Unprocessed\nPseudogene"))) %>%
  # 6.3.5. Compute the proportion of each gene type
  group_by(NAME) %>% 
  mutate(total = n()) %>%
  group_by(NAME, gene_type) %>% 
  summarise(prop = 100*n()/unique(total)) %>%
  ungroup() %>%
  mutate(NAME = factor(NAME, levels = c("FS-LA-4c", 'FS-LA-6c', 'FS-LA-8c', 'FS-LA-10c', 'FS-LA-12c', 'FS-19c'))) %>%
  # 6.3.6. Visualize results
  ggplot(aes(y = prop, x = NAME, fill = gene_type)) + 
  geom_hline(yintercept = seq(0,100,010), linetype = "dashed", color = "darkgrey") +
  geom_bar(stat = "identity", color = "black") +
  theme_cowplot(font_size = 18) + 
  theme(legend.text = element_text(size = 14),
        axis.text.x = element_text(angle = 45, hjust = 1), 
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank()) +
  scale_fill_manual(values = mycolors) +
  scale_y_continuous(breaks = seq(0,100,10)) +
  xlab("") +
  ylab("Expressed Genes (%)") 


ggsave(p.LA.content, bg = "white",
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_LA_content.pdf", dpi = 300, width = 8, height = 6)

# 7. Impact of Tn5 on the results

# 7.1. Select all the HEK LA data
mydata <- med.exon %>%
  filter(!grepl("262_HEK_LA_12c_01_B3|247_HEK_LV1_D7|247_HEK_LV1_B6", ID)) %>% 
  filter(PLATE %in% c("309", "247", "230", "260", "261", "262", "258")) %>%
  mutate(NAME = case_when(
    PLATE %in% c("247") & grepl("LV1", ID) | PLATE == "309" & dCTP == "3.3" ~ "FS-19c\n250 nL",
    PLATE == "230" & CYCLES == "8" & grepl(ID, pattern = "00625") ~ "FS-LA-8c\n62.5 nL",
    PLATE == "230" & CYCLES == "8" & grepl(ID, pattern = "0125") ~ "FS-LA-8c\n125 nL",
    PLATE == "230" & CYCLES == "8" & grepl(ID, pattern = "0015625") ~ "FS-LA-8c\n1.56 nL",
    PLATE == "260" & CYCLES == "6" & grepl(ID, pattern = "00078125") ~ "FS-LA-6c\n7.81 nL",
    PLATE == "260" & CYCLES == "6" & grepl(ID, pattern = "0015625") ~ "FS-LA-6c\n15.63 nL",
    PLATE == "258" & CYCLES == "4" & grepl(ID, pattern = "00039063") ~ "FS-LA-4c\n3.91 nL",
    PLATE == "258" & CYCLES == "4" & grepl(ID, pattern = "0015625") ~ "FS-LA-4c\n15.63 nL",
    PLATE == "258" & CYCLES == "4" & grepl(ID, pattern = "0078125") ~ "FS-LA-4c\n78.13 nL",
    PLATE == "261" & CYCLES == "10" & grepl(ID, pattern = "0015625") ~ "FS-LA-10c\n15.63 nL",
    PLATE == "261" & CYCLES == "10" & grepl(ID, pattern = "00625") ~ "FS-LA-10c\n62.5 nL",
    PLATE == "262" & CYCLES == "12" & grepl(ID, pattern = "01") ~ "FS-LA-12c\n100 nL",
    PLATE == "262" & CYCLES == "12" & grepl(ID, pattern = "003125") ~ "FS-LA-12c\n31.25 nL")) %>%
  filter(!is.na(NAME), DEPTH == 250000) %>%
  left_join(resqc, by = "ID") %>%
  mutate(TN5 = ifelse(is.na(TN5), 0.250, TN5))
write.table(mydata$ID, "/home/vincent.hahaut/Desktop/IDs.txt", col.names = F, row.names = F, quote = F, append = TRUE)

# 7.2. For each PCR cycle condition

# Create three graphics: 
# Gene Counts
# Read tag distribution
# Mapping Statistics
myplots <- list()
for(cycle in c(4,6,8,10,12)){
  
  mydata.tmp <- filter(mydata, CYCLES %in% c(cycle, 19))
  
  p.signif <- rstatix::dunn_test(data = mydata.tmp, count_0 ~ NAME, p.adjust.method = "bonferroni") %>%
  filter(p.adj < 0.05) %>% 
  rstatix::add_y_position(fun = "max", scales = "free") %>%
  mutate(p.adj.signif = formatC(p.adj, format = "e", digits = 2))

  p.gene <- ggplot() +
    geom_hline(yintercept = seq(0,12000,1000), linetype = "dotted", color = "darkgrey") +
    geom_violin(data = mydata.tmp, aes(y = count_0, x = NAME, fill = reorder(NAME, count_0, median)), width = 0.65) +
    geom_boxplot(data = mydata.tmp, aes(y = count_0, x = NAME, fill = reorder(NAME, count_0, median)), width = 0.1) +
    scale_fill_manual(values = c("FS-19c\n250 nL" = "#B10026",
                                 "FS-LA-8c\n62.5 nL" = "#41B6C4", 
                                 "FS-LA-8c\n125 nL" = "#225EA8", 
                                 "FS-LA-8c\n1.56 nL" = "#EDF8B1", 
                                 "FS-LA-6c\n7.81 nL" = "#EDF8B1", 
                                 "FS-LA-6c\n15.63 nL" = "#41B6C4", 
                                 "FS-LA-4c\n3.91 nL" = "#EDF8B1",
                                 "FS-LA-4c\n15.63 nL" = "#41B6C4",
                                 "FS-LA-4c\n78.13 nL" = "#225EA8",
                                 "FS-LA-10c\n15.63 nL" = "#EDF8B1", 
                                 "FS-LA-10c\n62.5 nL" = "#41B6C4", 
                                 "FS-LA-12c\n100 nL" = "#41B6C4",
                                 "FS-LA-12c\n31.25 nL" = "#EDF8B1")) +
    scale_y_continuous(breaks = seq(0,12000,1000), label = scales::comma) +
    geom_text(data = mydata.tmp, aes(x = NAME, label=paste0("n = ", ..count..)), y = -150, stat='count', size=5) +
    ylab("Genes Detected") + 
    xlab("") +
    theme_cowplot(font_size = 20) +
    theme(legend.position = "none",
          legend.title = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1)) +
    ggpubr::stat_pvalue_manual(p.signif, label = "p = {p.adj.signif}",  tip.length = 0.01, step.increase = 0.04, size = 4)

  
  p.resqc <- filter(resqc, ID %in% mydata.tmp$ID) %>%
    select(ID, UTR3_Exons, UTR5_Exons, CDS_Exons, Introns, TES_down_10kb, TSS_up_10kb, totalReadTags) %>%
    mutate(Intergenic = totalReadTags - UTR3_Exons - UTR5_Exons - CDS_Exons - Introns - TES_down_10kb - TSS_up_10kb) %>%
    gather(key, val, -ID, -totalReadTags) %>%
    mutate(perc = 100*val/totalReadTags) %>%
    left_join(mydata.tmp %>% select(NAME, ID), by = "ID") %>%
    mutate(key = factor(as.factor(key), levels = c("TES_down_10kb", "TSS_up_10kb", "UTR3_Exons", "UTR5_Exons", "Intergenic", "Introns", "CDS_Exons"))) %>%
    filter(!key %in% c("TES_down_10kb", "TSS_up_10kb")) %>%
    ggplot(aes(y = NAME, x = perc, fill = NAME)) +
    geom_violin() +
    geom_boxplot(width = 0.1) +
    theme_cowplot(font_size = 20) +
    facet_wrap("key", scales = "free_x", nrow = 1) +
    scale_fill_manual(values = c("FS-19c\n250 nL" = "#B10026",
                                 "FS-LA-8c\n62.5 nL" = "#41B6C4", 
                                 "FS-LA-8c\n125 nL" = "#225EA8", 
                                 "FS-LA-8c\n1.56 nL" = "#EDF8B1", 
                                 "FS-LA-6c\n7.81 nL" = "#EDF8B1", 
                                 "FS-LA-6c\n15.63 nL" = "#41B6C4", 
                                 "FS-LA-4c\n3.91 nL" = "#EDF8B1",
                                 "FS-LA-4c\n15.63 nL" = "#41B6C4",
                                 "FS-LA-4c\n78.13 nL" = "#225EA8",
                                 "FS-LA-10c\n15.63 nL" = "#EDF8B1", 
                                 "FS-LA-10c\n62.5 nL" = "#41B6C4", 
                                 "FS-LA-12c\n100 nL" = "#41B6C4",
                                 "FS-LA-12c\n31.25 nL" = "#EDF8B1")) +
    theme(legend.position = "none",
          rect = element_rect(size = 0.75, color = "black"), 
          axis.text.x = element_text(angle = 45, hjust = 1),
          panel.border = element_rect(size = 0.75, color = "black"), 
          panel.grid.major.x = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
          axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
    ylab("") +
    xlab("Read Tags (%)") 
  
  
  p.star <- select(mydata.tmp, NAME, P_uniquely_mapped_reads, P_multiple_Loci, P_too_short_read) %>%
    gather(key, val, -NAME) %>%
    mutate(key = case_when(key == "P_uniquely_mapped_reads" ~ "Uniquely Mapped", 
                           key == "P_multiple_Loci" ~ "Multi-Mapped", 
                           key == "P_too_short_read" ~ "Unmapped"),
           key = factor(as.factor(key), levels = c("Uniquely Mapped", "Multi-Mapped", "Unmapped")),
           FILL = paste(key, NAME)) %>%
    mutate(val = as.numeric(as.character(val))) %>%
    ggplot(aes(x = NAME, y = val, fill = NAME)) +
    geom_violin() +
    geom_boxplot(width = 0.1) +
    ylab("Reads (%)") +
    xlab("") +
    facet_wrap("key", nrow = 1, scales = "free_y") +
    theme_cowplot(font_size = 20) +
    theme(legend.position = "none",
          rect = element_rect(size = 0.75, color = "black"), 
          panel.border = element_rect(size = 0.75, color = "black"), 
          axis.text.x = element_text(size = 16, angle = 45, hjust = 1),
          panel.grid.major.x = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
          axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
    scale_fill_manual(values = c("FS-19c\n250 nL" = "#B10026",
                                 "FS-LA-8c\n62.5 nL" = "#41B6C4", 
                                 "FS-LA-8c\n125 nL" = "#225EA8", 
                                 "FS-LA-8c\n1.56 nL" = "#EDF8B1", 
                                 "FS-LA-6c\n7.81 nL" = "#EDF8B1", 
                                 "FS-LA-6c\n15.63 nL" = "#41B6C4", 
                                 "FS-LA-4c\n3.91 nL" = "#EDF8B1",
                                 "FS-LA-4c\n15.63 nL" = "#41B6C4",
                                 "FS-LA-4c\n78.13 nL" = "#225EA8",
                                 "FS-LA-10c\n15.63 nL" = "#EDF8B1", 
                                 "FS-LA-10c\n62.5 nL" = "#41B6C4", 
                                 "FS-LA-12c\n100 nL" = "#41B6C4",
                                 "FS-LA-12c\n31.25 nL" = "#EDF8B1"))
  myplots[[cycle]] <- plot_grid(p.gene, p.resqc, p.star, nrow = 1, rel_widths = c(0.18, 0.44, 0.38))
  
  ggsave(plot_grid(p.gene, p.resqc, p.star, nrow = 1, labels = "AUTO", rel_widths = c(0.18, 0.44, 0.38)), bg = "white",
         filename = paste0("/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_cycles_", cycle, ".pdf"),
         dpi = 450, width = 30, height = 6)
  
}

ggsave(plot_grid(myplots[[4]],
                 myplots[[6]],
                 myplots[[8]],
                 myplots[[10]],
                 myplots[[12]], nrow = 5, labels = "auto", label_size = 30), bg = "white", 
       filename = paste0("/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_cycles_ALLCYCLES_ALLTN5.pdf"),
       dpi = 380, width = 36, height = 32)

```

# CELL-TO-CELL CORRELATION LOW AMP

## Cell-to-Cell Correlation

```{r main correlation, eval = FALSE, echo = F, include = FALSE, warning = FALSE, comment = F, message = F, cache = FALSE, fig.align="center}

# 1. Select the data of interest
mydata <- med.exon %>%
  filter(PLATE %in% c("309", "247", "230", "260", "261", "262", "258")) %>%
  # Known outliers (doublets)
  filter(!grepl("262_HEK_LA_12c_01_B3|247_HEK_LV1_D7|247_HEK_LV1_B6", ID)) %>% 
  mutate(NAME = case_when(
    PLATE %in% c("247") & grepl("LV1", ID) | PLATE == "309" & dCTP == "3.3" ~ "FS-19c",
    # PLATE == "230" & CYCLES == "8" & grepl(ID, pattern = "00625") ~ "FS-8c",
    # PLATE == "230" & CYCLES == "8" & grepl(ID, pattern = "0125") ~ "8c_0.125",
    PLATE == "230" & CYCLES == "8" & grepl(ID, pattern = "0015625") ~ "FS-LA-8c",
    PLATE == "260" & CYCLES == "6" & grepl(ID, pattern = "0015625") ~ "FS-LA-6c",
    # PLATE == "260" & CYCLES == "6" & grepl(ID, pattern = "00078125") ~ "FS-6c\nTn5:15.625nL",
    # PLATE == "258" & CYCLES == "4" & grepl(ID, pattern = "00039063") ~ "4c_0.0039063",
    PLATE == "258" & CYCLES == "4" & grepl(ID, pattern = "0015625") ~ "FS-LA-4c",
    # PLATE == "258" & CYCLES == "4" & grepl(ID, pattern = "0078125") ~ "4c_0.078125",
    # PLATE == "261" & CYCLES == "10" & grepl(ID, pattern = "0015625") ~ "10c_0.015625",
    PLATE == "261" & CYCLES == "10" & grepl(ID, pattern = "00625") ~ "FS-LA-10c",
    PLATE == "262" & CYCLES == "12" & grepl(ID, pattern = "01") ~ "FS-LA-12c"
    # PLATE == "262" & CYCLES == "12" & grepl(ID, pattern = "003125") ~ "12c_0.03125"
    )) %>%
  filter(!is.na(NAME), DEPTH == 250000) %>%
  mutate(NAME = factor(as.factor(NAME), levels = c("FS-19c", "FS-LA-12c", "FS-LA-10c", "FS-LA-8c", "FS-LA-6c", "FS-LA-4c")),
         sce.ids = str_replace(paste0("X", ID, "_250000_1"), "-", "."))
write.table(mydata$ID, "/home/vincent.hahaut/Desktop/IDs.txt", col.names = F, row.names = F, quote = F, append = TRUE)
         
# 2. Expressed Genes

# 2.1. Select cells of interest
sce.expressed <- sce[,colnames(sce) %in% mydata$sce.ids]

# 2.2. Add Groups
sce.expressed$GROUP <- mydata$NAME[match(colnames(sce.expressed), mydata$sce.ids)]

# 2.3. Get genes expressed in all groups with >0 read in >2 cells
expressed.genes <- lapply(unique(sce.expressed$GROUP), function(x) row.names(
  sce.expressed[,sce.expressed$GROUP == x])[rowSums(counts(sce.expressed[,sce.expressed$GROUP == x]) > 0) > 2])
expressed.genes <- Reduce(union, expressed.genes)

sce.expressed <- sce.expressed[row.names(sce.expressed) %in% expressed.genes,]

# SAVE MATRIX
newNames <- read_tsv("/home/vincent.hahaut/Desktop/fastqTransform.txt")
newNames$OLD_FASTQ_ID <- str_replace_all(paste0("X", newNames$OLD_FASTQ_ID, "_250000_1"), "-", ".")

colnames(sce.expressed) <- newNames$FINAL_FASTQ_ID[match(colnames(sce.expressed), newNames$OLD_FASTQ_ID)]

write_rds(sce.expressed, "/home/vincent.hahaut/data_storage/transfer/Mendeley/HEK_FS_lowAmplification_250K.rds")

# 3. Correlations
corrs <- data_frame()
for(j in unique(c("FS-19c", "FS-LA-12c", "FS-LA-10c", "FS-LA-4c", "FS-LA-8c", "FS-LA-6c"))){
    sce.tmp <- counts(sce.expressed[,sce.expressed$GROUP == j])
    p <- pcaPP::cor.fk(sce.tmp) 
    
    corrs <- bind_rows(corrs,
                       data_frame(GROUP = j,
                                  tau = p[upper.tri(p)]))
}

# 4. Statistical Comparisons
shapiro.test(corrs$tau)

p.signif <-   corrs %>%
  rstatix::dunn_test(tau ~ GROUP, p.adjust.method = "bonferroni") %>%
  filter(p.adj < 0.05) %>% 
  rstatix::add_y_position(fun = "max", scales = "free") %>%
  mutate(p.adj.signif = formatC(p.adj, format = "e", digits = 2))

# 5. Visualize results
p6 <- corrs %>%
  mutate(GROUP = factor(GROUP, levels = c("FS-19c", "FS-LA-12c", "FS-LA-10c", "FS-LA-8c", "FS-LA-6c", "FS-LA-4c"))) %>%
  ggplot(aes(x = GROUP, y = tau)) + 
  geom_hline(yintercept = seq(0.45, 0.7, 0.05), linetype = "dotted", color = "darkgrey") +
  scale_y_continuous(breaks = seq(0.4,0.7,0.1)) +
  geom_violin(aes(fill = GROUP)) +
  geom_boxplot(aes(fill = GROUP), width = 0.1) +
  theme_cowplot(font_size = 24) +
    theme(legend.position = "none",
        axis.text.x = element_text(size = 18, angle = 45, hjust = 1), 
        axis.title = element_text(size = 20)) +
  xlab("") +
  scale_fill_manual(values = c("FS-LA-4c" = "#FED976", 
"FS-LA-8c" = "#FD8D3C", 
"FS-LA-6c" = "#FEB24C", 
"FS-LA-10c" = "#FC4E2A", 
"FS-LA-12c" = "#E31A1C", 
"FS-19c" = "#B10026")) +
  ylab("Cell-to-Cell Correlation (tau)") +
    ggpubr::stat_pvalue_manual(p.signif, label = "p = {p.adj.signif}",  step.increase = 0.025, tip.length = 0.01, size = 4) 

ggsave(p6, 
       filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/HEK_LA_kendallCorr.pdf", bg = "white", dpi = 450, width = 6, height = 8)
```
