---
title: "UMI in R2"
author: "V. Hahaut, D. Pavlinic & S. Picelli."
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r setup, include=FALSE}

source("/home/vincent.hahaut/Desktop/FLASH-Seq/R/general/4_aggregate_files_functions.R")
options(scipen=999)

# LIBRARIES / FUNCTIONS
library(tidyverse)
library(cowplot)
```

# TSO position in reads

```{r TSO position in reads - paired-ends, include=FALSE}

# GOAL: Extract the start position of the UMI in read 1 and read 2

if(!file.exists("/home/vincent.hahaut/data_storage/FS_intermediate_file/TSO_motifs.pos.rds")){
  
  suppressPackageStartupMessages(library(doSNOW))
  suppressPackageStartupMessages(library(doParallel))
  threads <- 10
    
  cl <- makeCluster(threads)
  registerDoSNOW(cl)
  clusterCall(cl, function() library(tidyverse))
    
  ## Where does the cut occur ?
  
  # 1. FASTQ Paths
  R1 <-list.files("/home/vincent.hahaut/data_storage/ORGANOIDS/FASTQ_DEEP/", full.names = TRUE, pattern = "R1")
  R1.ss3 <-list.files("/home/vincent.hahaut/data_storage/ORGANOIDS/FASTQ_MMU/", full.names = TRUE, pattern = "R1")

  R1 <- c(R1[1:50], R1.ss3[1:50])
  R2 <- str_replace(R1, "R1", "R2")
  IDs <- str_split(R1, "/", simplify = T)[,8]

  mymotifs <- data_frame()
  for(i in 1:length(R1)){
    
    message(i)
    # 2. Read File
    fq.R1 <- ShortRead::sread(ShortRead::readFastq(R1[i]))
    fq.R2 <- ShortRead::sread(ShortRead::readFastq(R2[i]))

    # 3. Extract the TSO position in the read (prior to UMI extraction)
    # In order to keep in simple, expect perfect matching

    # 3.1. Define the Pattern to search for
    # SS3 is a more complicated case as it we can't exactly rely on any spacer sequence to localise the UMI in case of cut
    # The detection is therefore less reliable
    pattern = case_when(
      grepl("5_SS3|6_SS3|SS3fwd|385_|SS3_fwd", IDs[i]) ~ "GCAATG........GG",
      grepl("STRToligoT_Spacer2|FSoligoT_Spacer2", IDs[i]) ~ "GT........ATAACGG",
      grepl("STRToligoT_Spacer1|FSoligoT_Spacer1|SS3oligoT_Spacer1", IDs[i]) ~ "GT........CAGCAGG",
      grepl("AAGCA", IDs[i]) ~ "GT........AAGCAGG",
      grepl("CATCA", IDs[i]) ~ "GT........CATCAGG",
      grepl("CTGAC", IDs[i]) ~ "GT........ATGACGG",
      grepl("ATGAC", IDs[i]) ~ "GT........ATGACGG",
      grepl("CTAAC|382_|316_|315_|317_|318_", IDs[i]) ~ "GT........CTAACGG",
      grepl("STRToligoT_Fstso|FSoligoT_FStso|SS3oligoT_Fstso", IDs[i]) ~ "AGAGT........GG")
    
    # 3.2. Extract the position
    starts.R1 <- regexpr(pattern, fq.R1)
    starts.R2 <- regexpr(pattern, fq.R2)
    starts.R1 <- starts.R1[!starts.R1 == -1]
    starts.R2 <- starts.R2[!starts.R2 == -1]
    
    # 3.3. Combine results and get the percentage of UMI starting at each position
    res <- bind_rows(data_frame(ID = IDs[i],
                                     starts = starts.R1,
                                     TYPE = "R1",
                                     pattern.match = length(starts.R1)),
                          data_frame(ID = IDs[i],
                                     starts = starts.R2,
                                     TYPE = "R2",
                                     pattern.match = length(starts.R2))) %>%
      group_by(ID, TYPE, starts, pattern.match) %>%
      summarise(n = n(), 
                perc = 100*n()/unique(pattern.match))

    mymotifs <- bind_rows(mymotifs, res)

  }
  stopCluster(cl)

  # 4. Aggregate results
  motifs <- bind_rows(mymotifs) %>%
    mutate(GROUP = case_when(grepl("5_SS3|6_SS3|385_|SS3_fwd", ID) ~ "SS3",
             grepl("STRToligoT_Spacer2", ID) ~ "TSO-ATAAC\nSTRT-dT",
             grepl("STRToligoT_Spacer1", ID) ~ "TSO-CAGCA\nSTRT-dT",
             grepl("AAGCA", ID) ~ "TSO-AAGCA\nSTRT-dT",
             grepl("CATCA", ID) ~ "TSO-CATCA\nSTRT-dT",
             grepl("CTGAC", ID) ~ "TSO-CTGAC\nSTRT-dT",
             grepl("ATGAC", ID) ~ "TSO-ATGAC\nSTRT-dT",
             grepl("CTAAC|382_|316_", ID) ~ "TSO-CTAAC\nSTRT-dT",
             grepl("LA_TSOspacer1", ID) ~ "LA-TSO-CAGCA\nSTRT-dT",
             grepl("STRToligoT_Fstso", ID) ~ "TSO-FS-UMI\nSTRT-dT",
             grepl("FSoligoT_Spacer2", ID) ~ "TSO-ATAAC\nFS-dT",
             grepl("FSoligoT_Spacer1", ID) ~ "TSO-CAGCA\nFS-dT",
             grepl("FSoligoT_FStso", ID) ~ "TSO-FS-UMI\nFS-dT",
             grepl("SS3oligoT_Spacer1", ID) ~ "TSO-CAGCA\nSS3-dT",
             grepl("SS3oligoT_Fstso", ID) ~ "TSO-FS-UMI\nSS3-dT"))
  
  write_rds(motifs, "/home/vincent.hahaut/data_storage/FS_intermediate_file/TSO_motifs.pos.rds")
} else {
  motifs <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/TSO_motifs.pos.rds")
}

# 5. Resynchronise position of UMI based on method for visualization purpose
# The start position is the one of the selected TSO motif
# For visualization purporse extract the UMI position
motifs <- motifs %>% 
  mutate(start.synchro = case_when(
    GROUP == "SS3" ~ starts + 6,
    GROUP %in% c("STRToligoT_Fstso, FSoligoT_FStso, SS3oligoT_Fstso") ~ starts + 5,
    TRUE  ~ starts + 2),
    FACET = paste0("Read: ", TYPE, "\nGROUP: ", GROUP))  %>% 
  filter(start.synchro < 25) %>%
  mutate(FACET = factor(FACET, levels = c("Read: R1\nGROUP: TSO-CTAAC\nSTRT-dT", "Read: R2\nGROUP: TSO-CTAAC\nSTRT-dT",  "Read: R1\nGROUP: SS3",  "Read: R2\nGROUP: SS3")))

# 6. Visualization
p.motifs.pos <- ggplot() + 
  geom_boxplot(data = motifs, aes(y = as.numeric(perc), x = start.synchro, fill = as.factor(start.synchro))) +
  xlab("UMI Start Position (0-base)") +
  ylab("Reads (%)") +
  facet_wrap("FACET", nrow = 4, scales = "free_x")+
  theme_bw(base_size = 20) +
  theme(axis.text.x = element_text(size = 12, angle = 90, hjust = 1), ) +
  ylim(c(0,100)) +
  scale_x_continuous(breaks = 1:24, limits = c(1,25)) +
  theme(legend.position = "none")

ggsave(p.motifs.pos, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/UMIposition.pdf", width = 10, height = 13, dpi = 450)
```

# Gene-body coverage

```{r get Coverage, include=FALSE}

# 1. Read the gene body coverages from ReSQC
paths.ReSQC.cov <- makePath(seqPaths = "/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2/", suffix = "ReSQC/ID_UMIinR1_genebody.geneBodyCoverage.txt")
coverage.R1 <- getRESeQC_coverage(paths = paths.ReSQC.cov$paths, sample.ids = paths.ReSQC.cov$sample.ids, threads = 10) 

paths.ReSQC.cov <- makePath(seqPaths = "/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2/", suffix = "ReSQC/ID_UMIinR2_genebody.geneBodyCoverage.txt")
coverage.R2 <- getRESeQC_coverage(paths = paths.ReSQC.cov$paths, sample.ids = paths.ReSQC.cov$sample.ids, threads = 10) 

# 2. Aggregate results
# filter out samples with <10K mapped reads
mycover <- bind_rows(
  coverage.R2 %>% mutate(TYPE = "R2"),
  coverage.R1 %>% mutate(TYPE = "R1")) %>%
  mutate(NAME = ifelse(grepl("316", ID), "FS-UMI\nCTAAC\nRetinal Organoids", "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735")) %>%
  distinct() %>%
  gather(Percentile, reads, -ID, -NAME, -TYPE) %>%
  group_by(ID) %>%
  mutate( total = sum(reads)) %>%
  ungroup() %>%
  filter(total > 10000) %>%
  group_by(ID, TYPE) %>%
  mutate(percentage = 100*reads / sum(reads)) %>%
  group_by(TYPE, NAME, Percentile) %>%
  summarise(mean.c = mean(percentage),
            sd.c = sd(percentage)) %>%
  ungroup() 

# 3. Visualize results
p.cover <- ggplot(mycover, aes(x = as.numeric(Percentile), color = TYPE, fill = TYPE)) +
  geom_ribbon(aes(ymin = mean.c - sd.c, ymax = mean.c + sd.c), colour = NA) +
  geom_line(aes(y=mean.c), size = 1.25) +
  facet_wrap("NAME") +
  theme_bw(base_size = 20) +
  scale_y_continuous(breaks = 0:7) +
  scale_color_manual(values = c("R1" = "#d12c1f",
                                "R2" = "#074f94")) + 
  scale_fill_manual(values = c("R1" = adjustcolor("#d12c1f", alpha = 0.25),
                                "R2" = adjustcolor("#074f94", alpha = 0.25))) +
  ylab("Coverage (%)") +
  xlab("Gene Body Percentile (5'->3')") +
  theme(legend.title = element_blank(),
        axis.text = element_text(size = 22),
        axis.title = element_text(size = 24),
        legend.background = element_blank(),
        legend.justification = c(1, 1), legend.position = c(0.95, 0.95))
  
ggsave(p.cover, filename = paste0("/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/UMIinR2.Coverage.pdf"), bg = "white", dpi = 450, height = 6, width = 12)
```


# Mapping Statistics

```{r stats, include=FALSE}

# 1. Load the files

paths.STAR.log <- makePath(seqPaths = "/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2/", suffix = "STAR/ID_UMIinR1_downsampled_Log.final.out")
STAR.R1.log <- getSTARLog(paths = paths.STAR.log$paths, sample.ids = paths.STAR.log$sample.ids, threads = 10)
paths.STAR.log <- makePath(seqPaths = "/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2/", suffix = "STAR/ID_UMIinR2_downsampled_Log.final.out")
STAR.R2.log <- getSTARLog(paths = paths.STAR.log$paths, sample.ids = paths.STAR.log$sample.ids, threads = 10) %>%
  filter(grepl("316", ID))

# 2. Aggregate the results
# Remove low quality cells (unmapped < 25)
df.star <- bind_rows(STAR.R1.log %>% mutate(TYPE = "R1", GROUP = ifelse(grepl("316", ID), "FS-UMI\nCTAAC\nRetinal Organoids", "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735")), 
                     STAR.R2.log %>% mutate(TYPE = "R2", GROUP = ifelse(grepl("316", ID), "FS-UMI\nCTAAC\nRetinal Organoids", "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735"))) %>%
  filter(as.numeric(P_too_short_read) < 25) %>%
  dplyr::select(ID, P_uniquely_mapped_reads, P_too_short_read, P_multiple_Loci, TYPE, GROUP) %>%
  gather(reads, perc, -ID, -TYPE, -GROUP) %>%
  mutate(perc = as.numeric(perc),
             reads = case_when(
               reads == "P_uniquely_mapped_reads" ~ "Uniquely\nMapped",
               reads == "P_too_short_read" ~ "Unmapped",
               reads == "P_multiple_Loci" ~ "Multiple\nLoci"),
             reads = factor(reads, levels = c("Uniquely\nMapped", "Multiple\nLoci", "Unmapped")))

# 3. Compare the UMI in R1/R2 using a wilcoxon-test
p.signif <- df.star %>%
  filter(GROUP == "FS-UMI\nCTAAC\nRetinal Organoids") %>%
  group_by(reads, GROUP) %>%
  rstatix::wilcox_test(perc~TYPE, p.adjust.method = "bonferroni") %>%
  filter(p < 0.05) %>% 
  rstatix::add_y_position(fun = "max", scales = "free") %>%
  mutate(p.adj.signif = formatC(p, format = "e", digits = 2))


# 4. Visualize results

# 4.1. FS-UMI
p.star <- ggplot() +
  geom_violin(data = df.star %>% filter(GROUP == "FS-UMI\nCTAAC\nRetinal Organoids"), aes(y = perc, x = TYPE, fill = reads)) +
  geom_boxplot(data = df.star %>% filter(GROUP == "FS-UMI\nCTAAC\nRetinal Organoids"), aes(y = perc, x = TYPE, fill = reads), width = 0.1) +
  facet_wrap("reads")+
  scale_fill_manual(values = c("Uniquely\nMapped" = "#E41A1C", 
                                 "Multiple\nLoci" =  "#377EB8", 
                                 "Unmapped" =  "#4DAF4A")) +
  ylab("Reads (%)") + 
  xlab("") +
  theme_cowplot(font_size = 18) +
  theme(legend.position = "none", 
        strip.background = element_rect(size = 0.75, color = "black"), 
        panel.border = element_rect(size = 0.75, color = "black"), 
        axis.text.x = element_text(size = 16),
        strip.text = element_text(face = "bold"),
        panel.grid.major.y = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
        axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
  scale_y_continuous(breaks = seq(0,100,10), limits = c(0,101)) +
  ggpubr::stat_pvalue_manual(p.signif, label = "p={p.adj.signif}",  tip.length = 0.01, step.increase = 0.04, size = 5) + ggtitle("FS-UMI CTAAC")

# 4.2. SS3
p.star.ss3 <- ggplot() +
  geom_violin(data = df.star %>% filter(GROUP == "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735") %>% mutate(facet = "SS3\n"), aes(y = perc, x = TYPE, fill = reads)) +
  geom_boxplot(data = df.star %>% filter(GROUP == "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735") %>% mutate(facet = "SS3\n"), aes(y = perc, x = TYPE, fill = reads), width = 0.1, position = position_dodge(0.9)) +
  scale_fill_manual(values = c("Uniquely\nMapped" = "#E41A1C", 
                                 "Multiple\nLoci" =  "#377EB8", 
                                 "Unmapped" =  "#4DAF4A")) +
  facet_wrap("facet") +
  ylab("Percentage Reads") + 
  xlab("") +
  theme_cowplot(font_size = 18) +
  scale_y_continuous(breaks = seq(0,100,10), limits = c(0,101)) +
  theme(legend.position = "none", 
        strip.background = element_rect(size = 0.75, color = "black"), 
        panel.border = element_rect(size = 0.75, color = "black"), 
        axis.text.x = element_text(size = 12),
        strip.text = element_text(face = "bold"),
        panel.grid.major.y = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
        axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) + ggtitle("SS3")

ggsave(p.star, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/UMIinR2.star.stats.pdf", dpi = 450, bg = "white", width = 7, height = 4)
```

# Read Distribution

```{r Read distribution, include=FALSE}

# 1. Load resuts
paths.ReSQC.distr <- makePath(seqPaths = "/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2/", suffix = "ReSQC/ID_UMIinR1_readDistribution.txt")
ReSQC.R1.distr <- getRESeQC_readDistribution(paths = paths.ReSQC.distr$paths, sample.ids = paths.ReSQC.distr$sample.ids, threads = 10)
paths.ReSQC.distr <- makePath(seqPaths = "/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2/", suffix = "ReSQC/ID_UMIinR2_readDistribution.txt")
ReSQC.R2.distr <- getRESeQC_readDistribution(paths = paths.ReSQC.distr$paths, sample.ids = paths.ReSQC.distr$sample.ids, threads = 10) %>%
  filter(grepl("316", ID))

# 2. Aggregate
# filter out samples with <10K read tags
df.readDistrib <- bind_rows(
  ReSQC.R1.distr %>%  mutate(TYPE = "R1", GROUP = ifelse(grepl("316", ID), "FS-UMI\nCTAAC\nRetinal Organoids", "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735")), 
  ReSQC.R2.distr %>%  mutate(TYPE = "R2", GROUP = ifelse(grepl("316", ID), "FS-UMI\nCTAAC\nRetinal Organoids", "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735"))) %>%
    gather(distribution, val, -ID, -TYPE, -GROUP) %>%
    filter(!distribution %in% c("TSS_up_1kb", "TSS_up_5kb", "TSS_down_1kb", "TSS_down_1kb",  "TES_down_10kb", "TES_down_1kb" , "TES_down_5kb", "TSS_up_10kb", "totalReadTags")) %>%
    mutate(distribution = case_when(distribution == "UTR5_Exons" ~ "5'UTR\nExons",
                                    distribution == "UTR3_Exons" ~ "3'UTR\nExons",
                                    distribution == "CDS_Exons" ~ "CDS\nExons",
                                    distribution == "Introns" ~ "Introns",
                                    distribution == "Intergenic" ~ "Intergenic")) %>%
  group_by(ID) %>%
  mutate(n = sum(val)) %>% 
  ungroup() %>%
  filter(n > 10000) %>% 
  group_by(ID, TYPE) %>%
  mutate(total = sum(val)) %>%
  ungroup() %>%
  mutate(perc = 100*val / total) 

# 3. Compare the % of read tags in UMI in R1 / R2 using a wilcoxon test
p.signif <- df.readDistrib %>%
  filter(GROUP == "FS-UMI\nCTAAC\nRetinal Organoids") %>%
  group_by(distribution) %>%
  rstatix::wilcox_test(perc~TYPE, p.adjust.method = "bonferroni") %>%
  filter(p < 0.05) %>% 
  rstatix::add_y_position(fun = "max", scales = "free") %>%
  mutate(p.adj.signif = formatC(p, format = "e", digits = 2))

# 4. Visualize results

# 4.1. FS-UMI
p.readDistrib <- ggplot() +
    geom_violin(data = df.readDistrib %>% filter(GROUP != "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735"), aes(y = perc, x = TYPE, fill = distribution)) +
    geom_boxplot(data = df.readDistrib %>% filter(GROUP != "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735"), aes(y = perc, x = TYPE, fill = distribution), width = 0.1) +
    scale_fill_manual(values = c("CDS\nExons" = "#1B9E77", 
                                 "Introns" =  "#D95F02", 
                                 "5'UTR\nExons" =  "#7570B3",
                                 "3'UTR\nExons" =  "#E7298A",
                                 "Intergenic" =  "#66A61E")) +
    facet_wrap("distribution", nrow = 1) +
    ylab("Reads (%)") + 
    xlab("") +
    theme_cowplot(font_size = 18) +
    scale_y_continuous(breaks = seq(0,80,10), limits = c(0,83)) +
    theme(legend.position = "none", 
          strip.background = element_rect(size = 0.75, color = "black"), 
          panel.border = element_rect(size = 0.75, color = "black"), 
          axis.text.x = element_text(size = 16),
          strip.text = element_text(face = "bold"),
          panel.grid.major.y = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
          axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) +
  ggpubr::stat_pvalue_manual(p.signif, label = "p={p.adj.signif}",  tip.length = 0.01, step.increase = 0.04, size = 5) 


# 4.2. SS3 data
p.readDistrib.SS3 <- ggplot() +
    geom_violin(data = df.readDistrib %>% filter(GROUP != "FS-UMI\nCTAAC\nRetinal Organoids"), aes(y = perc, x = distribution, fill = distribution)) +
    geom_boxplot(data = df.readDistrib %>% filter(GROUP != "FS-UMI\nCTAAC\nRetinal Organoids"), aes(y = perc, x = distribution, fill = distribution), width = 0.05, position = position_dodge(0.95)) +
  scale_y_continuous(breaks = seq(0,80,10), limits = c(0,81.5)) +
    scale_fill_manual(values = c("CDS\nExons" = "#1B9E77", 
                                 "Introns" =  "#D95F02", 
                                 "5'UTR\nExons" =  "#7570B3",
                                 "3'UTR\nExons" =  "#E7298A",
                                 "Intergenic" =  "#66A61E")) +
    facet_wrap("GROUP") +
    ylab("Percentage Reads") + 
    xlab("") +
    theme_cowplot(font_size = 18) +
    theme(legend.position = "none", 
          strip.background = element_rect(size = 0.75, color = "black"), 
          panel.border = element_rect(size = 0.75, color = "black"), 
          axis.text.x = element_text(size = 12),
          strip.text = element_text(face = "bold"),
          panel.grid.major.y = element_line(linetype = "dotted", color = "darkgrey", size = 0.5),
          axis.line.y.right = element_line(color = "black", linetype = "solid", size = 1)) 

ggsave(p.readDistrib, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/UMIinR2.resqc.stats.pdf", dpi = 450, bg = "white", width = 10, height = 4)

```

# Number of detected genes

```{r Number of detected genes, include=FALSE}

# 1. Get the gene counts
# at 30K raw UMI reads
ids <- list.files("/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2/", recursive = FALSE, full.names = FALSE)
paths.R1 <- paste0("/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2/", ids, "/FEATURECOUNTS/", ids, ".umi.UMIinR1.30000.tsv.gz")
paths.R2 <- paste0("/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2/", ids, "/FEATURECOUNTS/", ids, ".umi.UMIinR2.30000.tsv.gz")
paths.R1 <- paths.R1[file.exists(paths.R1)]
paths.R2 <- paths.R2[file.exists(paths.R2)]
ids.R1 <- str_split(paths.R1, "/", simplify = T)[,7]
ids.R2 <- str_split(paths.R2, "/", simplify = T)[,7]

counts.R1 <- getUMICounts(paths.R1, ids.R1)
counts.R2 <- getUMICounts(paths.R2, ids.R2)

# 2. Get the number of detected genes per cell
# 3 different read thresholds
med.R1 <- lapply(c(0,1,5), function(x) colSums(counts(counts.R1) > x))
med.R2 <- lapply(c(0,1,5), function(x) colSums(counts(counts.R2) > x))

# 3. Aggregate
med <- bind_rows(
  bind_cols(med.R1) %>% mutate(ID = names(med.R1[[1]]), TYPE = "R1"),
  bind_cols(med.R2) %>% mutate(ID = names(med.R2[[1]]), TYPE = "R2")) 
colnames(med) <- c("count_0", "count_1", "count_5", "ID", "TYPE")
med <- pivot_longer(med, cols = c(count_0, count_1, count_5), names_to = "threshold", values_to = "counts") %>%
  mutate(GROUP = ifelse(grepl("316", ID), "FS-UMI\nCTAAC\nRetinal Organoids", "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735"),
         threshold = case_when(threshold == "count_0" ~ ">0 read",
                               threshold == "count_1" ~ ">1 read",
                               threshold == "count_5" ~ ">5 reads"))

# 4. Visualize results
p.signif <- med %>% 
  filter(GROUP == "FS-UMI\nCTAAC\nRetinal Organoids") %>%
  group_by(threshold) %>%
  rstatix::wilcox_test(counts~TYPE, p.adjust.method = "bonferroni") %>%
  filter(p < 0.05) %>% 
  rstatix::add_y_position(fun = "max", scales = "free") %>%
  mutate(p.adj.signif = formatC(p, format = "e", digits = 2))

# 4.1. FS-UMI
p.counts.umi <- ggplot() + 
  geom_violin(data = med %>% filter(GROUP == "FS-UMI\nCTAAC\nRetinal Organoids"), aes(x = threshold, y = counts, fill = TYPE)) +
  geom_boxplot(data = med %>% filter(GROUP == "FS-UMI\nCTAAC\nRetinal Organoids"), aes(x = threshold, y = counts, fill = TYPE), width = 0.1, position = position_dodge(0.9)) +
  theme_bw(base_size = 20) +
  xlab("") +
  scale_fill_brewer(palette = "Set1") +
  ylab("Genes Detected")  +
  theme(legend.justification = c(1, 1), 
        legend.title = element_blank(),
        legend.position = c(0.95, 0.90), 
        legend.background = element_blank()) +
  scale_y_continuous(breaks = seq(0,4500,500), labels = scales::comma, limits = c(0,5000)) +
  ggtitle("FS-UMI CTAAC") +
  geom_text(aes(x = c(0.75, 1.25), y = c(0,0), label = c(n = "n = 380", "n = 213")), size = 4) 

# 4.2. SS3
p.counts.umi.SS3 <- ggplot() + 
  geom_violin(data = med %>% filter(GROUP != "FS-UMI\nCTAAC\nRetinal Organoids", TYPE == "R1"), aes(x = threshold, y = counts, fill = TYPE)) +
  geom_boxplot(data = med %>% filter(GROUP != "FS-UMI\nCTAAC\nRetinal Organoids", TYPE == "R1"), aes(x = threshold, y = counts, fill = TYPE), width = 0.05, position = position_dodge(0.95)) +
  theme_bw(base_size = 20) +
  theme(legend.position = "none", axis.text.x = element_text(size = 9)) + 
  xlab("") +
  scale_fill_brewer(palette = "Set1") +
  scale_y_continuous(breaks = seq(0,3000,500), limits = c(0,3500), labels = scales::comma) +
  ylab("Genes Detected") + ggtitle("SS3") +
  geom_text(aes(x = 1, y = 0, label = "n = 365"), size = 4) 


ggsave(p.counts.umi, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/UMIinR2.counts.pdf", dpi = 450, bg = "white", width = 8, height = 5)

```

# Cell-to-cell correlations

```{r Cell-to-cell correlations, include=FALSE}

# 1. Correlations

# 1.1. Get samples with >30K UMI reads in both R1 and R2
IDinCommon <- intersect(colnames(counts.R1), colnames(counts.R2)) 
IDinCommon <- IDinCommon[grepl("316", IDinCommon)]

# 1.2. Calculate the correlation between gene counts (pair R1-R2 same cell)
suppressPackageStartupMessages(library(doSNOW))
suppressPackageStartupMessages(library(doParallel))
cl <- makeCluster(10)
registerDoSNOW(cl)
clusterCall(cl, function() library(tidyverse))
clusterCall(cl, function() library(SingleCellExperiment))

corrs <- foreach(i = 1:length(IDinCommon)) %dopar% {
  j <- IDinCommon[i]
  mycounts <- left_join(
    counts(counts.R1[,colnames(counts.R1) == j]) %>% 
      rownames_to_column("gene"),
    counts(counts.R2[,colnames(counts.R2) == j]) %>% 
      rownames_to_column("gene"), by = "gene") 
  
  mycounts[is.na(mycounts)] <- 0
  
  p <- cor.test(mycounts[,2], mycounts[,3], method = "kendall")
  
  return(data_frame(ID = j,
                    tau = broom::tidy(p)$estimate))
}
stopCluster(cl)

# 4.3. Aggregate
mycorrs <- bind_rows(corrs)

# 4.4. Visualization
p.corrs <- ggplot(mycorrs, aes(x = "", y = tau, fill = "col")) + 
  geom_hline(yintercept = seq(0.7,0.9,0.05), linetype = "dashed", color = "darkgrey") +
  geom_violin() + 
  geom_boxplot(width = 0.1) + 
  theme_cowplot(font_size = 20) +
  scale_fill_manual(values = "#ffbc98") +
  theme(legend.position = "none") +
  ylab("R1/R2 UMI gene count\ncorrelations (tau)") +
  xlab("")

ggsave(p.corrs, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/UMIinR2.corrs.pdf", dpi = 450, bg = "white", width = 3, height = 5)

```

# ARCHIVE 

# CODE WORKING - USED IN DATA EXPLORATIION BUT NOT IN THE FINAL MANUSCRIPT

```{r TSO position in reads - single-ends, include=FALSE}

# 1. TSO position in read from HEK 293T cells (single-end reads)
if(!file.exists("/home/vincent.hahaut/data_storage/FS_intermediate_file/TSO_motifs.pos.singleEnd.rds")){
  
  suppressPackageStartupMessages(library(doSNOW))
  suppressPackageStartupMessages(library(doParallel))
  threads <- 10
    
  cl <- makeCluster(threads)
  registerDoSNOW(cl)
  clusterCall(cl, function() library(tidyverse))
    
  ## Where does the cut occurs ?
  
  # 1.1. Read FASTQ R1 files
  R1 <-list.files("/home/vincent.hahaut/data_storage/210322_NB551561_0044_AHNTL2AFX2/out/", full.names = TRUE, pattern = "_R1_")
  IDs <- str_replace(list.files("/home/vincent.hahaut/data_storage/210322_NB551561_0044_AHNTL2AFX2/out/", full.names = FALSE, pattern = "_R1_"), pattern = "_S[0-9]+_R1_001.fastq.gz", "")
  
  R1 <- R1[!grepl("276_LA_TSO|277_TSO_SS3|Undetermined", R1)]
  IDs <- IDs[!grepl("276_LA_TSO|277_TSO_SS3|Undetermined", IDs)]
  
  mymotifs <- foreach(i = 1:length(R1)) %dopar% {
    
    # 1.2. Read File
    fq.R1 <- ShortRead::sread(ShortRead::readFastq(R1[i]))

    # 1.3. Extract the TSO position in the read (prior to UMI extraction)
    # In order to keep it simple, expect perfect matching

    # 1.3.1. Define the Pattern to search for
    # SS3 is a more complicated case as it we can't exactly rely on any spacer sequence to localize the UMI in case of cut
    # The call is therefore less reliable
    pattern = case_when(
      grepl("5_SS3|6_SS3|SS3fwd|385_", IDs[i]) ~ "GCAATG........GG",
      grepl("STRToligoT_Spacer2|FSoligoT_Spacer2", IDs[i]) ~ "GT........ATAACGG",
      grepl("SS3-CGTAC", IDs[i]) ~ "TG........CGTACGG",
      grepl("STRToligoT_Spacer1|FSoligoT_Spacer1|SS3oligoT_Spacer1", IDs[i]) ~ "GT........CAGCAGG",
      grepl("AAGCA", IDs[i]) ~ "GT........AAGCAGG",
      grepl("CATCA", IDs[i]) ~ "GT........CATCAGG",
      grepl("CTGAC", IDs[i]) ~ "GT........ATGACGG",
      grepl("ATGAC", IDs[i]) ~ "GT........ATGACGG",
      grepl("CTAAC|382_", IDs[i]) ~ "GT........CTAACGG",
      grepl("STRToligoT_Fstso|FSoligoT_FStso|SS3oligoT_Fstso", IDs[i]) ~ "AGAGT........GG")
    
    # 1.3.2. Find the start position
    starts.R1 <- regexpr(pattern, fq.R1)
    starts.R1 <- starts.R1[!starts.R1 == -1]

    return(bind_rows(data_frame(ID = IDs[i],
                                     starts = starts.R1,
                                     TYPE = "R1",
                                     pattern.match = length(starts.R1))) %>%
      group_by(ID, TYPE, starts, pattern.match) %>%
      summarise(n = n(), 
                perc = 100*n()/unique(pattern.match)))
  }
  stopCluster(cl)

  # 1.4. Aggregate the results
  motifs <- bind_rows(mymotifs) %>%
    mutate(GROUP = case_when(grepl("5_SS3|6_SS3|385_", ID) ~ "SS3",
             grepl("SS3fwd", ID) ~ "SS3\nHagemann-J.",
             grepl("STRToligoT_Spacer2", ID) ~ "TSO-ATAAC\nSTRT-dT",
             grepl("STRToligoT_Spacer1", ID) ~ "TSO-CAGCA\nSTRT-dT",
             grepl("AAGCA", ID) ~ "TSO-AAGCA\nSTRT-dT",
             grepl("CATCA", ID) ~ "TSO-CATCA\nSTRT-dT",
             grepl("CTGAC", ID) ~ "TSO-CTGAC\nSTRT-dT",
             grepl("ATGAC", ID) ~ "TSO-ATGAC\nSTRT-dT",
             grepl("CTAAC|382_", ID) ~ "TSO-CTAAC\nSTRT-dT",
             grepl("SS3-CGTAC", ID) ~ "SS3-CGTAC\nSS3-dT",
             grepl("LA_TSOspacer1", ID) ~ "LA-TSO-CAGCA\nSTRT-dT",
             grepl("STRToligoT_Fstso", ID) ~ "TSO-FS-UMI\nSTRT-dT",
             grepl("FSoligoT_Spacer2", ID) ~ "TSO-ATAAC\nFS-dT",
             grepl("FSoligoT_Spacer1", ID) ~ "TSO-CAGCA\nFS-dT",
             grepl("FSoligoT_FStso", ID) ~ "TSO-FS-UMI\nFS-dT",
             grepl("SS3oligoT_Spacer1", ID) ~ "TSO-CAGCA\nSS3-dT",
             grepl("SS3oligoT_Fstso", ID) ~ "TSO-FS-UMI\nSS3-dT"))
  
  write_rds(motifs, "/home/vincent.hahaut/data_storage/FS_intermediate_file/TSO_motifs.pos.singleEnd.rds")
} else {
  motifs <- read_rds("/home/vincent.hahaut/data_storage/FS_intermediate_file/TSO_motifs.pos.singleEnd.rds")
}

# 1.5. Visualization
motifs <- motifs %>% 
  mutate(start.synchro = case_when(
    GROUP == "SS3" ~ starts + 6,
    GROUP %in% c("STRToligoT_Fstso, FSoligoT_FStso, SS3oligoT_Fstso") ~ starts + 5,
    TRUE  ~ starts + 2))  %>% 
  filter(start.synchro < 25)

p.motifs.pos <- ggplot() + 
  geom_boxplot(data = motifs, aes(y = as.numeric(perc), x = start.synchro, fill = as.factor(start.synchro))) +
  xlab("UMI Start Position (0-base)") +
  ylab("Reads (%)") +
  facet_wrap("GROUP", ncol = 3)+
  theme_bw(base_size = 20) +
  theme(axis.text.x = element_text(size = 12, angle = 90, hjust = 1), ) +
  ylim(c(0,100)) +
  scale_x_continuous(breaks = 1:24, limits = c(1,25)) +
  theme(legend.position = "none") 

ggsave(p.motifs.pos, filename = "/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/UMIposition.singleEnd.pdf", width = 12, height = 14, dpi = 300)
```

```{r duplication rate and number of UMI reads, include=FALSE}

# 1. Calculate the Duplication % of R1/R2 UMI
suppressPackageStartupMessages(library(doSNOW))
suppressPackageStartupMessages(library(doParallel))
threads <- 5
    
cl <- makeCluster(threads)
registerDoSNOW(cl)
clusterCall(cl, function() library(tidyverse))
    
  
# 1.1. Read FASTQ R1 files
R1 <-list.files("/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2", full.names = TRUE, recursive = T, pattern = "umi.UMIinR1.R1.fq.gz")
R2 <-list.files("/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2", full.names = TRUE, recursive = T, pattern = "umi.UMIinR2.R1.fq.gz")
all.reads <-list.files("/home/vincent.hahaut/data_storage/ORGANOIDS/UMIinR2", full.names = TRUE, recursive = T, pattern = "sample.R1.fastq")
ID.R1 <- str_split(R1, pattern = "/", simplify = T)[,7]
ID.R2 <- str_split(R2, pattern = "/", simplify = T)[,7]

umi.R1 <- umi.R2 <- list()
stats <- foreach(i = 1:length(R1)) %dopar% {
  
  # 1.2. Read File
  R1.id <- ShortRead::id(ShortRead::readFastq(R1[i]))
  R2.id <- ShortRead::id(ShortRead::readFastq(R2[i]))
  nReads <- length(ShortRead::readFastq(all.reads[i]))
  # differentiate early cuts
  R1.l <- ShortRead::width(ShortRead::readFastq(R1[i]))
  R2.l <- ShortRead::width(ShortRead::readFastq(R2[i]))
  
  # 1.3. UMI
  R1.id <- str_replace_all(str_extract(pattern = "_........ ", string = R1.id), "_| ", "")
  R2.id <- str_replace_all(str_extract(pattern = "_........ ", string = R2.id), "_| ", "")

  R1.trimmed <- R1.id[R1.l >= 70]
  R1.full <- R1.id[R1.l < 70]

  # 1.4. % duplication
  data_frame(ID = ID.R1[i],
             percentage_duplicated_R1.trimmed = 100*length(unique(R1.trimmed)) / length(R1.trimmed),
             percentage_duplicated_R1.full = 100*length(unique(R1.full)) / length(R1.full),
             percentage_duplicated_R2 = 100*length(unique(R2.id)) / length(R2.id),
             n_UMI_R2 = length(R2.id),
             n_UMI_R1 = length(R1.id),
             n_UMI_R1.trimmed = length(R1.full),
             n_UMI_R1.full = length(R1.trimmed),
             percentage_UMIinR1 = 100*length(R1.id)/nReads,
             percentage_UMIinR2 = 100*length(R2.id)/nReads,
             total.reads = nReads) %>%
    return()
}
stopCluster(cl)

# 1.5. Aggregate
stats.aggregated <- bind_rows(stats) %>%
  filter(n_UMI_R1 > 2000) %>%
  mutate(GROUP = ifelse(grepl("316", ID), "FS-UMI\nCTAAC\nRetinal Organoids", "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735"),
         cumultated = percentage_UMIinR1 + percentage_UMIinR2)

# 1.6. Visualization
percentageUMI <- plot_grid(nrow = 1,
  ggplot(stats.aggregated, aes(x = GROUP, y = percentage_UMIinR1, fill = GROUP)) + 
    geom_hline(yintercept = seq(0,90, 10), linetype = "dashed", color = "darkgrey") +
    geom_violin() +
    geom_boxplot(width = 0.1) + 
    theme_cowplot() +
    theme(legend.position = "none") +
    xlab("") + 
    ylab("UMI Reads (%)") +
    scale_y_continuous(breaks = seq(0,90,10), limits = c(0,100)) +
    scale_fill_manual(values = c("FS-UMI\nCTAAC\nRetinal Organoids" = "#FF850D", "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735" = "#4DAF4A")) +
    ggtitle("UMI in R1"),
  ggplot(stats.aggregated, aes(x = GROUP, y = percentage_UMIinR2, fill = GROUP)) + 
    geom_hline(yintercept = seq(0,90, 10), linetype = "dashed", color = "darkgrey") +
    geom_violin() +
    theme_cowplot() + 
    theme(legend.position = "none") +
    scale_y_continuous(breaks = seq(0,90,10), limits = c(0,100)) +
    geom_boxplot(width = 0.1) + 
    xlab("") + 
    ylab("UMI Reads (%)") +
    ggtitle("UMI in R2") +
    scale_fill_manual(values = c("FS-UMI\nCTAAC\nRetinal Organoids" = "#FF850D", "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735" = "#4DAF4A")),
  ggplot(stats.aggregated, aes(x = GROUP, y = cumultated, fill = GROUP)) + 
    geom_hline(yintercept = seq(0,90, 10), linetype = "dashed", color = "darkgrey") +
    geom_violin() +
    theme_cowplot() + 
    theme(legend.position = "none") +
    scale_y_continuous(breaks = seq(0,90,10), limits = c(0,100)) +
    geom_boxplot(width = 0.1) + 
    xlab("") + 
    ylab("UMI Reads (%)") +
    ggtitle("Cumulated R1/R2") +
    scale_fill_manual(values = c("FS-UMI\nCTAAC\nRetinal Organoids" = "#FF850D", "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735" = "#4DAF4A"))
)

ggsave(percentageUMI, filename="/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/ORGANOIDS_percentageUMI.pdf", bg = "white", dpi = 450, width = 12, height = 5)

# 4.2. Duplication
dups <- stats.aggregated %>%
  select(GROUP, percentage_duplicated_R1.full, percentage_duplicated_R1.trimmed, percentage_duplicated_R2) %>%
  filter(GROUP != "SS3\nMice Fibroblast\nHagemann-J. et al\nE-MTAB-8735") %>%
  pivot_longer(cols = c(percentage_duplicated_R1.full, percentage_duplicated_R1.trimmed, percentage_duplicated_R2)) %>%
  mutate(name = case_when(name == "percentage_duplicated_R1.full" ~ "R1 UMI\nFull", 
                          name == "percentage_duplicated_R1.trimmed" ~ "R1 UMI\nTn5 Cut", 
                          name == "percentage_duplicated_R2" ~ "R2 UMI")) 
  
p.sign <- ggpubr::compare_means(data = dups, value ~  name) %>% 
  mutate(y = c(93,97,101),
         p.signif = ifelse(p.signif == "****", "***", p.signif),
         x = match(group1, levels(factor(dups$name))),
         xend = match(group2, levels(factor(dups$name))))

p.dups <- ggplot() + 
    geom_hline(yintercept = seq(0,90, 10), linetype = "dashed", color = "darkgrey") +
    geom_violin(data = dups, aes(x = name, y = value, fill = name)) +
    geom_boxplot(data = dups, aes(x = name, y = value, fill = name), width = 0.1) + 
    theme_cowplot(font_size = 20) + 
    theme(legend.position = "none") +
    scale_y_continuous(breaks = seq(0,90,10), limits = c(0,102)) +
    xlab("") + 
    ylab("Percentage Unique UMI") +
    scale_fill_brewer(palette = "Set1") +
  geom_segment(data = p.sign, aes(x = x, xend = xend, y = y, yend = y)) +
  geom_text(data = p.sign, aes(x = (x+xend)/2, y = y+0.35, label = p.signif), size = 6) 

    
ggsave(p.dups, filename="/home/vincent.hahaut/data_storage/FS_intermediate_file/FIGURES/ORGANOIDS_percentageDups.pdf", bg = "white", dpi = 450, width = 6, height = 5)
```
